// @dalbit-yaksok/core/node/class/core.js
import { NotDefinedIdentifierError } from "../../error/variable.js";
import { ObjectValue, ValueType } from "../../value/base.js";
export class ClassValue extends ValueType {
  name;
  body;
  definitionScope;
  parentClass;
  static friendlyName = '클래스';
  constructor(name, body, definitionScope, parentClass){
    super(), this.name = name, this.body = body, this.definitionScope = definitionScope, this.parentClass = parentClass;
  }
  toPrint() {
    return `<클래스 ${this.name}>`;
  }
}
export class InstanceValue extends ObjectValue {
  className;
  static friendlyName = '인스턴스';
  scope;
  memberLookupRootScope;
  classValue;
  constructor(className){
    super(), this.className = className;
  }
  toPrint() {
    return `<${this.className} 인스턴스>`;
  }
}
export class SuperValue extends ObjectValue {
  instance;
  scope;
  static friendlyName = '상위';
  constructor(instance, scope){
    super(), this.instance = instance, this.scope = scope;
  }
  toPrint() {
    return '<상위>';
  }
}
export function getInheritanceChain(classValue) {
  const chain = [];
  let cursor = classValue;
  while(cursor){
    chain.unshift(cursor);
    cursor = cursor.parentClass;
  }
  return chain;
}
export function resolveClassValueFromInstance(scope, instance) {
  if (instance.classValue) {
    return instance.classValue;
  }
  try {
    const value = scope.getVariable(instance.className);
    if (value instanceof ClassValue) {
      return value;
    }
    return undefined;
  } catch (error) {
    if (error instanceof NotDefinedIdentifierError) {
      return undefined;
    }
    throw error;
  }
}
