// @dalbit-yaksok/core/node/class/declare-validation.ts
import {
    AlreadyDefinedMemberFunctionError,
    ConstructorArityAmbiguousError,
} from '../../error/class.js'
import type { YaksokError } from '../../error/common.ts'
import type { Token } from '../../prepare/tokenize/token.ts'
import { DeclareFunction } from '../function.js'
import type { ClassValue } from './core.ts'
import {
    getDuplicatedConstructorArities,
    isConstructorDeclaration,
} from './constructor.js'

export function validateDuplicatedConstructorArity(
    className: string,
    classValue: ClassValue,
    tokens: Token[],
): YaksokError[] {
    return getDuplicatedConstructorArities(classValue).map(
        (arity) =>
            new ConstructorArityAmbiguousError({
                resource: {
                    className,
                    arity,
                },
                tokens,
            }),
    )
}

export function validateDuplicatedMemberFunctionName(
    className: string,
    bodyChildren: unknown[],
    tokens: Token[],
): YaksokError[] {
    const functionCountByName = new Map<string, number>()

    for (const child of bodyChildren) {
        if (!(child instanceof DeclareFunction)) continue
        if (isConstructorDeclaration(child)) continue

        functionCountByName.set(
            child.name,
            (functionCountByName.get(child.name) ?? 0) + 1,
        )
    }

    const duplicatedNames = [...functionCountByName.entries()]
        .filter(([, count]) => count > 1)
        .map(([name]) => name)

    return duplicatedNames.map(
        (functionName) =>
            new AlreadyDefinedMemberFunctionError({
                resource: {
                    className,
                    functionName,
                },
                tokens,
            }),
    )
}
