// @dalbit-yaksok/core/node/class/member-runtime.js
import { DotAccessOnlyOnInstanceError } from "../../error/class.js";
import { NotDefinedIdentifierError } from "../../error/variable.js";
import { InstanceValue, SuperValue } from "./core.js";
export function resolveMemberAccessTarget(target, tokens) {
  if (target instanceof InstanceValue) {
    return {
      scope: target.scope,
      instance: target
    };
  }
  if (target instanceof SuperValue) {
    return {
      scope: target.scope,
      instance: target.instance
    };
  }
  throw new DotAccessOnlyOnInstanceError({
    tokens
  });
}
export function findMemberFunction(scope, instance, functionName) {
  let cursor = scope;
  while(cursor){
    const fromCurrentScope = cursor.functions.get(functionName);
    if (fromCurrentScope) {
      return fromCurrentScope;
    }
    if (cursor === instance.memberLookupRootScope) {
      break;
    }
    cursor = cursor.parent;
  }
  return undefined;
}
export function findVariableOwnerScopeInMemberChain(scope, instance, memberName) {
  let cursor = scope;
  while(cursor){
    if (memberName in cursor.variables) {
      return cursor;
    }
    if (cursor === instance.memberLookupRootScope) {
      break;
    }
    cursor = cursor.parent;
  }
  return undefined;
}
export function getMemberVariable(scope, instance, memberName, tokens) {
  const owner = findVariableOwnerScopeInMemberChain(scope, instance, memberName);
  if (!owner) {
    const error = new NotDefinedIdentifierError({
      resource: {
        name: memberName
      }
    });
    error.tokens = tokens;
    throw error;
  }
  return owner.getVariable(memberName, tokens);
}
export function setMemberVariable(scope, instance, memberName, value, tokens) {
  const owner = findVariableOwnerScopeInMemberChain(scope, instance, memberName);
  if (owner) {
    owner.setLocalVariable(memberName, value, tokens);
    return;
  }
  scope.setLocalVariable(memberName, value, tokens);
}
