// @dalbit-yaksok/core/node/class/new-instance-constructor.js
import { ConstructorArityAmbiguousError, ConstructorArityMismatchError } from "../../error/class.js";
import { createConstructorLayersFromInheritanceChain, getDeclaredConstructorsInClass, resolveConstructorByArity } from "./constructor.js";
export function pickConstructorByArity(props) {
  const { layerScopes, arity, className, tokens } = props;
  const resolution = resolveConstructorByArity(layerScopes.map(({ klass })=>({
      className: klass.name,
      arities: getDeclaredConstructorsInClass(klass).map((constructor)=>constructor.arity)
    })), arity);
  if (resolution.kind === 'ambiguous') {
    throw new ConstructorArityAmbiguousError({
      resource: {
        className: resolution.className,
        arity
      },
      tokens
    });
  }
  for(let i = layerScopes.length - 1; i >= 0; i--){
    const { klass, scope } = layerScopes[i];
    const declaredConstructors = getDeclaredConstructorsInClass(klass);
    const matchingDeclaredConstructors = declaredConstructors.filter((constructor)=>constructor.arity === arity);
    for (const constructor of matchingDeclaredConstructors){
      const func = scope.functions.get(constructor.name);
      if (func) {
        return func;
      }
    }
  }
  if (resolution.kind === 'none') {
    return undefined;
  }
  const expectedArities = resolution.kind === 'mismatch' ? resolution.expectedArities : [
    arity
  ];
  throw new ConstructorArityMismatchError({
    resource: {
      className,
      expected: expectedArities,
      received: arity
    },
    tokens
  });
}
export function validateConstructorByArity(classValue, arity, tokens) {
  const resolution = resolveConstructorByArity(createConstructorLayersFromInheritanceChain(classValue), arity);
  if (resolution.kind === 'matched' || resolution.kind === 'none' || resolution.kind === 'ambiguous') {
    return [];
  }
  return [
    new ConstructorArityMismatchError({
      resource: {
        className: classValue.name,
        expected: resolution.expectedArities,
        received: arity
      },
      tokens
    })
  ];
}
