// @dalbit-yaksok/core/node/class/validation-analysis.js
import { CountLoop } from "../count-loop.js";
import { IfStatement } from "../IfStatement.js";
import { ListLoop } from "../listLoop.js";
import { Loop } from "../loop.js";
import { Block } from "../block.js";
import { BooleanLiteral } from "../primitive-literal.js";
import { Evaluable, Executable, Identifier } from "../base.js";
import { DeclareFunction } from "../function.js";
import { getInheritanceChain } from "./core.js";
function isSetMemberLikeNode(node) {
  return node instanceof Executable && '__kind' in node && node.__kind === 'SetMember' && 'target' in node && 'memberName' in node && typeof node.memberName === 'string';
}
function isSetVariableLikeNode(node) {
  return node instanceof Evaluable && '__kind' in node && node.__kind === 'SetVariable' && 'name' in node && typeof node.name === 'string';
}
function toStaticBoolean(condition) {
  if (!condition) return undefined;
  if (!(condition instanceof BooleanLiteral)) return undefined;
  return condition.toPrint() === '참';
}
export function collectGuaranteedMemberWritesFromBlock(block, options) {
  const writes = [];
  const collectFromNode = (node)=>{
    if (options.includePlainSetVariable && isSetVariableLikeNode(node)) {
      writes.push({
        name: node.name,
        target: 'self'
      });
      return;
    }
    if (isSetMemberLikeNode(node)) {
      const target = node.target;
      if (target instanceof Identifier && (target.value === '자신' || target.value === '상위')) {
        writes.push({
          name: node.memberName,
          target: target.value === '자신' ? 'self' : 'super'
        });
        return;
      }
    }
    if (node instanceof IfStatement) {
      for (const caseItem of node.cases){
        const staticBoolean = toStaticBoolean(caseItem.condition);
        if (staticBoolean === false) {
          continue;
        }
        if (staticBoolean === true) {
          writes.push(...collectGuaranteedMemberWritesFromBlock(caseItem.body, options));
        }
        return;
      }
      return;
    }
    if (node instanceof DeclareFunction) {
      return;
    }
    if (node instanceof Loop || node instanceof CountLoop || node instanceof ListLoop) {
      return;
    }
    if (node instanceof Block) {
      writes.push(...collectGuaranteedMemberWritesFromBlock(node, options));
    }
  };
  for (const child of block.children){
    collectFromNode(child);
  }
  return writes;
}
export function collectLikelyMemberNamesInClass(classValue) {
  const names = new Set();
  const collectSetMemberWritesInNode = (node)=>{
    if (isSetMemberLikeNode(node)) {
      const target = node.target;
      if (target instanceof Identifier && (target.value === '자신' || target.value === '상위')) {
        names.add(node.memberName);
        return;
      }
    }
    if (node instanceof Block) {
      for (const child of node.children){
        collectSetMemberWritesInNode(child);
      }
      return;
    }
    if (node instanceof IfStatement) {
      for (const caseItem of node.cases){
        collectSetMemberWritesInNode(caseItem.body);
      }
      return;
    }
    if (node instanceof Loop || node instanceof CountLoop || node instanceof ListLoop) {
      collectSetMemberWritesInNode(node.body);
    }
  };
  for (const klass of getInheritanceChain(classValue)){
    const deterministicWrites = collectGuaranteedMemberWritesFromBlock(klass.body, {
      includePlainSetVariable: true
    });
    for (const write of deterministicWrites){
      names.add(write.name);
    }
    for (const child of klass.body.children){
      if (!(child instanceof DeclareFunction)) continue;
      collectSetMemberWritesInNode(child.body);
    }
  }
  return names;
}
