// @pandacss/plugin-lightningcss/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  optimizeLightCss: () => optimizeLightCss,
  pluginLightningcss: () => pluginLightningcss
});
module.exports = __toCommonJS(index_exports);

// src/optimize-lightningcss.ts
var import_logger = require("@pandacss/logger");
var import_lightningcss = require("lightningcss");
var import_browserslist = __toESM(require("browserslist"));
var decoder = new TextDecoder();
function optimizeLightCss(code, options = {}) {
  const { minify = false, browserslist: targets } = options;
  const result = (0, import_lightningcss.transform)({
    // https://stackoverflow.com/questions/78790943/in-typescript-5-6-buffer-is-not-assignable-to-arraybufferview-or-uint8arr
    code: Buffer.from(code),
    minify,
    sourceMap: false,
    filename: "styles.css",
    include: import_lightningcss.Features.Nesting | import_lightningcss.Features.MediaRangeSyntax,
    targets: (0, import_lightningcss.browserslistToTargets)((0, import_browserslist.default)(targets)),
    errorRecovery: true
  });
  if (result.warnings.length) {
    const split = code.split("\n");
    import_logger.logger.warn(
      "css",
      result.warnings.map((w) => ({ ...w, line: split[w.loc.line - 1] }))
    );
  }
  return decoder.decode(result.code);
}

// src/index.ts
function pluginLightningcss() {
  return {
    name: "@pandacss/plugin-lightningcss",
    hooks: {
      "css:optimize": ({ css, minify, browserslist: browserslist2 }) => {
        return optimizeLightCss(css, { minify, browserslist: browserslist2 });
      }
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  optimizeLightCss,
  pluginLightningcss
});
