// @pandacss/plugin-lightningcss/dist/index.mjs
// src/optimize-lightningcss.ts
import { logger } from "@pandacss/logger";
import { Features, transform, browserslistToTargets } from "lightningcss";
import browserslist from "browserslist";
var decoder = new TextDecoder();
function optimizeLightCss(code, options = {}) {
  const { minify = false, browserslist: targets } = options;
  const result = transform({
    // https://stackoverflow.com/questions/78790943/in-typescript-5-6-buffer-is-not-assignable-to-arraybufferview-or-uint8arr
    code: Buffer.from(code),
    minify,
    sourceMap: false,
    filename: "styles.css",
    include: Features.Nesting | Features.MediaRangeSyntax,
    targets: browserslistToTargets(browserslist(targets)),
    errorRecovery: true
  });
  if (result.warnings.length) {
    const split = code.split("\n");
    logger.warn(
      "css",
      result.warnings.map((w) => ({ ...w, line: split[w.loc.line - 1] }))
    );
  }
  return decoder.decode(result.code);
}

// src/index.ts
function pluginLightningcss() {
  return {
    name: "@pandacss/plugin-lightningcss",
    hooks: {
      "css:optimize": ({ css, minify, browserslist: browserslist2 }) => {
        return optimizeLightCss(css, { minify, browserslist: browserslist2 });
      }
    }
  };
}
export {
  optimizeLightCss,
  pluginLightningcss
};
