// @dalbit-yaksok/core/util/get-function-declare-ranges.js
import { isClassStartingPattern, isFfiStartingPattern, isYaksokStartingPattern } from "./is-function-starting.js";
import { TOKEN_TYPE } from "../prepare/tokenize/token.js";
import { UnexpectedEndOfCodeError, UnexpectedTokenError } from "../error/prepare.js";
import { getEventDeclareRanges } from "./get-event-declare-ranges.js";
export function getFunctionDeclareRanges(tokens) {
  const yaksokFunctionDeclareRanges = getFunctionDeclareRangesByType(tokens, 'yaksok');
  const ffiFunctionDeclareRanges = getFunctionDeclareRangesByType(tokens, 'ffi');
  const classFunctionDeclareRanges = getFunctionDeclareRangesByType(tokens, 'class');
  const eventFunctionDeclareRanges = getEventDeclareRanges(tokens);
  return {
    yaksok: yaksokFunctionDeclareRanges,
    ffi: ffiFunctionDeclareRanges,
    class: classFunctionDeclareRanges,
    event: eventFunctionDeclareRanges
  };
}
function getFunctionDeclareRangesByType(_tokens, type) {
  const tokens = [
    ..._tokens
  ];
  let starterFinder = isYaksokStartingPattern;
  if (type === 'ffi') starterFinder = isFfiStartingPattern;
  if (type === 'class') starterFinder = isClassStartingPattern;
  const functionStartingIndexes = tokens.map(starterFinder).map((isStarting, index)=>isStarting ? index : -1).filter((index)=>index !== -1);
  const functionEndingIndexesByStartingIndex = functionStartingIndexes.map((startingIndex)=>getFunctionEndingIndex(tokens, startingIndex));
  const functionDeclareRanges = functionStartingIndexes.map((startingIndex, index)=>[
      startingIndex,
      functionEndingIndexesByStartingIndex[index]
    ]);
  if (type === 'yaksok' || type === 'class') {
    assertValidYaksokDeclare(tokens, functionDeclareRanges);
  } else {
    assertValidFfiDeclare(tokens, functionDeclareRanges);
  }
  return functionDeclareRanges;
}
function getFunctionEndingIndex(tokens, startingIndex) {
  const tokensFromStartingIndex = tokens.slice(startingIndex);
  const nearestNewLineIndexFromStart = tokensFromStartingIndex.findIndex((token)=>token.type === TOKEN_TYPE.NEW_LINE);
  if (nearestNewLineIndexFromStart === -1) {
    const lastToken = tokensFromStartingIndex[tokensFromStartingIndex.length - 1];
    throw new UnexpectedEndOfCodeError({
      resource: {
        expected: '줄넘김 이후에 함수의 코드'
      },
      position: lastToken.position
    });
  }
  return nearestNewLineIndexFromStart + startingIndex + 1;
}
function assertValidYaksokDeclare(tokens, functionDeclareRanges) {
  for (const [_, end] of functionDeclareRanges){
    const nextToken = tokens[end - 1];
    const nextNextToken = tokens[end];
    if (!nextNextToken) {
      throw new UnexpectedEndOfCodeError({
        resource: {
          expected: '들여쓰기'
        },
        position: nextToken.position
      });
    }
    if (nextNextToken.type !== TOKEN_TYPE.INDENT) {
      throw new UnexpectedTokenError({
        resource: {
          parts: '들여쓰기'
        },
        position: nextNextToken.position,
        tokens: [
          nextNextToken
        ]
      });
    }
  }
}
function assertValidFfiDeclare(tokens, functionDeclareRanges) {
  for (const [_, end] of functionDeclareRanges){
    const nextToken = tokens[end - 1];
    const nextNextToken = tokens[end];
    if (!nextNextToken) {
      throw new UnexpectedEndOfCodeError({
        resource: {
          expected: '번역 본문'
        },
        position: nextToken.position
      });
    }
    if (nextNextToken.type !== TOKEN_TYPE.FFI_BODY) {
      throw new UnexpectedTokenError({
        resource: {
          parts: '번역 본문'
        },
        position: nextNextToken.position,
        tokens: [
          nextNextToken
        ]
      });
    }
  }
}
