// @dalbit-yaksok/core/error/common.js
import { TOKEN_TYPE_TO_TEXT } from "../prepare/tokenize/token.js";
export class YaksokError extends Error {
  position;
  tokens;
  resource;
  codeFile;
  child;
  constructor(props){
    super();
    this.position = props.position;
    this.resource = props.resource;
    this.tokens = props.tokens;
  }
}
export function evaluableToText(evaluable) {
  let text = evaluable.constructor.friendlyName;
  try {
    text = bold(blue(evaluable.toPrint())) + dim(`(${text})`);
  } catch  {
  // If toPrint() is not implemented, ignore
  }
  return text;
}
export function valueTypeToText(valueType) {
  return bold(blue(valueType.toPrint())) + dim(`(${valueType.constructor.friendlyName})`);
}
export function operatorToText(operator) {
  let text = operator.constructor.friendlyName;
  const toPrint = operator.toPrint();
  if (toPrint !== 'unknown') text = blue(bold(toPrint)) + dim(`(${text})`);
  return text;
}
export function tokenToText(token) {
  const type = TOKEN_TYPE_TO_TEXT[token.type];
  const text = token.value.replace('\n', dim('줄바꿈'));
  return `${bold(`'${text}'`)}${dim(`(${type})`)}`;
}
export function expressionToText(node) {
  return `${bold(blue(node.toPrint()))}${dim(`(${node.constructor.friendlyName})`)}`;
}
export function bold(text) {
  return `\x1b[1m${text}\x1b[0m`;
}
export function blue(text) {
  return `\x1b[34m${text}\x1b[0m`;
}
export function dim(text) {
  return `\x1b[2m${text}\x1b[0m`;
}
export function underline(text) {
  return `\x1b[4m${text}\x1b[24m`;
}
