// @dalbit-yaksok/core/prepare/parse/dynamicRule/mention/create-mentioning-rules.ts
import { getTokensFromNodes } from '../../../../util/merge-tokens.js'
import { Mention, MentionScope } from '../../../../node/mention.js'
import { FunctionInvoke } from '../../../../node/function.js'
import { Identifier, Node } from '../../../../node/base.js'
import { Token } from '../../../tokenize/token.js'

import type { Rule } from '../../type.ts'

export function createMentioningRule(
    fileName: string,
    originalRule: Rule,
): Rule {
    const mergedPattern = [
        {
            type: Mention,
            value: fileName,
        },
        ...originalRule.pattern,
    ]

    return {
        pattern: mergedPattern,
        config: originalRule.config,
        flags: originalRule.flags,
        factory: createFactory(fileName, originalRule),
    }
}

function createFactory(fileName: string, rule: Rule) {
    return (nodes: Node[], tokens: Token[]) => {
        const childNodes = nodes.slice(1)
        const childTokens = getTokensFromNodes(childNodes)

        const child = rule.factory(nodes.slice(1), childTokens) as
            | Identifier
            | FunctionInvoke
        child.position = nodes[1].position

        return new MentionScope(fileName, child, tokens)
    }
}
