// @dalbit-yaksok/core/node/base.js
import { NotDefinedIdentifierError } from "../error/variable.js";
import { assertValidReturnValue } from "../util/assert-valid-return-value.js";
import { YaksokError } from "../error/common.js";
import { NotExecutableNodeError } from "../error/unknown-node.js";
import { AbortedSessionSignal } from "../executer/signals.js";
export class Node {
  tokens = [];
  static friendlyName = '노드';
  validate(_scope) {
    throw new Error(`${this.getNodeTypeName()} has no validate method`);
  }
  toJSON() {
    return {
      type: this.getNodeTypeName(),
      ...this
    };
  }
  toPrint() {
    throw new Error(`${this.getNodeTypeName()} has no toPrint method`);
  }
  getNodeTypeName() {
    return this.constructor.friendlyName || '노드';
  }
}
export class Executable extends Node {
  static friendlyName = '실행 가능한 노드';
  execute(_scope) {
    throw new Error(`${this.getNodeTypeName()} has no execute method`);
  }
  toPrint() {
    throw new Error(`${this.getNodeTypeName()} has no toPrint method`);
  }
  async onRunChild({ scope, childTokens, skipReport = false }) {
    if (scope.codeFile?.session?.stepUnit) {
      if (!(this instanceof scope.codeFile?.session?.stepUnit)) {
        return;
      }
    }
    if (scope.codeFile?.session?.signal?.aborted) {
      throw new AbortedSessionSignal(childTokens);
    }
    const executionDelay = scope.codeFile?.executionDelay ?? 0;
    if (executionDelay) {
      await new Promise((r)=>setTimeout(r, executionDelay));
    } else {
      await scope.codeFile?.session?.increaseTick();
    }
    // runningCode 이벤트를 resume 대기 후에 발생하도록 순서 변경
    if (scope.codeFile?.session?.paused || scope.codeFile?.session?.stepByStep) {
      await new Promise((resolve)=>{
        const unsubscribe = scope.codeFile?.session?.pubsub.sub('resume', ()=>{
          resolve(undefined);
          unsubscribe();
        });
      });
    }
    if (!skipReport && childTokens.length) {
      this.reportRunningCode(childTokens, scope);
    }
  }
  reportRunningCode(childTokens, scope) {
    const startPosition = childTokens[0].position;
    const endToken = childTokens[childTokens.length - 1];
    const endPosition = {
      line: endToken.position.line,
      column: endToken.position.column + endToken.value.length
    };
    scope.codeFile?.session?.pubsub.pub('runningCode', [
      {
        line: startPosition.line,
        column: startPosition.column
      },
      endPosition,
      scope,
      childTokens
    ]);
  }
}
export class Evaluable extends Executable {
  static friendlyName = '값이 있는 노드';
  execute(_scope) {
    throw new Error(`${this.getNodeTypeName()} has no execute method`);
  }
}
export class Identifier extends Evaluable {
  value;
  tokens;
  static friendlyName = '식별자';
  constructor(value, tokens){
    super(), this.value = value, this.tokens = tokens;
  }
  toPrint() {
    return this.value;
  }
  async execute(scope) {
    try {
      return scope.getVariable(this.value, this.tokens);
    } catch (e) {
      if (e instanceof NotDefinedIdentifierError) {
        try {
          const functionObject = scope.getFunctionObject(this.value);
          const functionResult = await functionObject.run({});
          assertValidReturnValue(this, functionResult);
          return functionResult;
        } catch (e2) {
          if (e2 instanceof NotDefinedIdentifierError) {
            e2.tokens = this.tokens;
          }
          throw e2;
        }
      }
      throw e;
    }
  }
  validate(scope) {
    try {
      scope.getVariable(this.value);
      return [];
    } catch (variableError) {
      if (!(variableError instanceof YaksokError)) {
        throw variableError;
      }
      if (!(variableError instanceof NotDefinedIdentifierError)) {
        variableError.tokens = this.tokens;
        return [
          variableError
        ];
      }
      try {
        scope.getFunctionObject(this.value);
        return [];
      } catch (functionError) {
        if (!(functionError instanceof YaksokError)) {
          throw functionError;
        }
      }
      variableError.tokens = this.tokens;
      return [
        variableError
      ];
    }
  }
}
export class Operator extends Node {
  value;
  tokens;
  static friendlyName = '연산자';
  constructor(value, tokens){
    super(), this.value = value, this.tokens = tokens;
  }
  toPrint() {
    return this.value ?? '알 수 없음';
  }
  async call(_left, _right) {
    throw new Error(`${this.getNodeTypeName()} has no call method`);
  }
  validate() {
    return [];
  }
}
export class Expression extends Node {
  value;
  tokens;
  static friendlyName = '표현식';
  constructor(value, tokens){
    super(), this.value = value, this.tokens = tokens;
  }
  toPrint() {
    return this.value;
  }
  validate() {
    const error = new NotExecutableNodeError({
      tokens: this.tokens,
      resource: {
        node: this
      }
    });
    return [
      error
    ];
  }
}
