// @dalbit-yaksok/core/node/block.ts
import { YaksokError } from '../error/common.js'
import { CannotParseError } from '../error/index.js'
import { Executable, type Node } from './base.js'
import { EOL } from './misc.js'

import type { Scope } from '../executer/scope.ts'
import type { Token } from '../prepare/tokenize/token.ts'

export class Block extends Executable {
    static override friendlyName = '코드 덩어리'

    children: Node[]

    constructor(
        content: Node[],
        public override tokens: Token[],
    ) {
        super()
        this.children = content
    }

    override async execute(scope: Scope): Promise<void> {
        for (const child of this.children) {
            if (child instanceof Executable) {
                if (scope.codeFile?.session?.canRunNode) {
                    if (
                        !(await scope.codeFile?.session?.canRunNode(
                            scope,
                            child,
                        ))
                    ) {
                        return
                    }
                }

                await this.onRunChild({
                    childTokens: child.tokens,
                    scope,
                })
                await child.execute(scope)
            } else if (child instanceof EOL) {
                continue
            } else {
                throw new CannotParseError({
                    resource: {
                        part: child,
                    },
                    tokens: child.tokens,
                })
            }
        }
    }

    override validate(scope: Scope): YaksokError[] {
        const childErrors = this.children
            .flatMap((child) => child.validate(scope))
            .filter((error) => error !== null)

        return childErrors
    }
}
