// @dalbit-yaksok/core/node/block.js
import { CannotParseError } from "../error/index.js";
import { Executable } from "./base.js";
import { EOL } from "./misc.js";
export class Block extends Executable {
  tokens;
  static friendlyName = '코드 덩어리';
  children;
  constructor(content, tokens){
    super(), this.tokens = tokens;
    this.children = content;
  }
  async execute(scope) {
    for (const child of this.children){
      if (child instanceof Executable) {
        if (scope.codeFile?.session?.canRunNode) {
          if (!await scope.codeFile?.session?.canRunNode(scope, child)) {
            return;
          }
        }
        await this.onRunChild({
          childTokens: child.tokens,
          scope
        });
        await child.execute(scope);
      } else if (child instanceof EOL) {
        continue;
      } else {
        throw new CannotParseError({
          resource: {
            part: child
          },
          tokens: child.tokens
        });
      }
    }
  }
  validate(scope) {
    const childErrors = this.children.flatMap((child)=>child.validate(scope)).filter((error)=>error !== null);
    return childErrors;
  }
}
