// @dalbit-yaksok/core/node/calculation.js
import { BooleanValue } from "../value/primitive.js";
import { Evaluable, Operator } from "./base.js";
import { DivideOperator, EqualOperator, GreaterThanOperator, GreaterThanOrEqualOperator, IntegerDivideOperator, LessThanOperator, LessThanOrEqualOperator, MinusOperator, ModularOperator, MultiplyOperator, NotEqualOperator, AndOperator, OrOperator, PlusOperator, PowerOperator, RangeOperator } from "./operator.js";
import { FormulaStackUnderflowError, InvalidFormulaError, NotBooleanTypeError } from "../error/calculation.js";
export class ValueWithParenthesis extends Evaluable {
  value;
  tokens;
  static friendlyName = '괄호로 묶인 값';
  constructor(value, tokens){
    super(), this.value = value, this.tokens = tokens;
  }
  async execute(scope) {
    return await this.value.execute(scope);
  }
  validate(scope) {
    return this.value.validate(scope);
  }
}
export class NotExpression extends Evaluable {
  value;
  tokens;
  static friendlyName = '부정';
  constructor(value, tokens){
    super(), this.value = value, this.tokens = tokens;
  }
  async execute(scope) {
    const value = await this.value.execute(scope);
    if (!(value instanceof BooleanValue)) {
      throw new NotBooleanTypeError({
        resource: {
          value
        },
        tokens: this.tokens
      });
    }
    return new BooleanValue(!value.value);
  }
  validate(scope) {
    return this.value.validate(scope);
  }
}
export class Formula extends Evaluable {
  terms;
  tokens;
  static friendlyName = '계산식';
  constructor(terms, tokens){
    super(), this.terms = terms, this.tokens = tokens;
  }
  async execute(scope) {
    const rpn = this.toRPN();
    const stack = [];
    for (const item of rpn){
      if (item instanceof Operator) {
        const right = stack.pop();
        const left = stack.pop();
        if (!left || !right) {
          throw new FormulaStackUnderflowError({
            resource: {
              formula: this
            },
            tokens: this.tokens
          });
        }
        const combinedTokens = [
          ...left.tokens,
          ...item.tokens,
          ...right.tokens
        ];
        // 연산자 결과도 thunk로 올림 — 피연산자가 먼저 평가된 후 combined 토큰 리포트
        const leftThunk = left.thunk;
        const rightThunk = right.thunk;
        stack.push({
          thunk: async ()=>{
            const result = await item.call(leftThunk, rightThunk);
            await this.onRunChild({
              scope,
              childTokens: combinedTokens
            });
            return result;
          },
          tokens: combinedTokens
        });
      } else {
        stack.push({
          thunk: async ()=>{
            await this.onRunChild({
              scope,
              childTokens: item.tokens
            });
            return item.execute(scope);
          },
          tokens: item.tokens
        });
      }
    }
    if (stack.length !== 1) {
      throw new InvalidFormulaError({
        resource: {
          formula: this
        },
        tokens: this.tokens
      });
    }
    return stack[0].thunk();
  }
  validate(scope) {
    return this.terms.filter((term)=>term instanceof Evaluable).flatMap((term)=>term.validate(scope));
  }
  toRPN() {
    const outputQueue = [];
    const operatorStack = [];
    for (const term of this.terms){
      if (term instanceof Operator) {
        while(operatorStack.length > 0 && this.shouldPopOperator(term, operatorStack[operatorStack.length - 1])){
          outputQueue.push(operatorStack.pop());
        }
        operatorStack.push(term);
      } else {
        outputQueue.push(term);
      }
    }
    while(operatorStack.length > 0){
      outputQueue.push(operatorStack.pop());
    }
    return outputQueue;
  }
  shouldPopOperator(current, top) {
    const currentPrecedence = this.getPrecedence(current);
    const topPrecedence = this.getPrecedence(top);
    if (current instanceof PowerOperator) {
      // Right associative: pop only if top has GREATER precedence
      return topPrecedence > currentPrecedence;
    }
    // Left associative: pop if top has GREATER OR EQUAL precedence
    return topPrecedence >= currentPrecedence;
  }
  getPrecedence(operator) {
    if (operator instanceof PowerOperator) return 5;
    if (operator instanceof MultiplyOperator || operator instanceof DivideOperator || operator instanceof IntegerDivideOperator || operator instanceof ModularOperator) {
      return 4;
    }
    if (operator instanceof PlusOperator || operator instanceof MinusOperator) {
      return 3;
    }
    if (operator instanceof GreaterThanOperator || operator instanceof LessThanOperator || operator instanceof GreaterThanOrEqualOperator || operator instanceof LessThanOrEqualOperator || operator instanceof EqualOperator || operator instanceof NotEqualOperator || operator instanceof RangeOperator) {
      return 2;
    }
    if (operator instanceof AndOperator) return 1;
    if (operator instanceof OrOperator) return 0;
    return -1;
  }
}
