// @dalbit-yaksok/core/node/conditional-loop.js
import { BreakSignal, ContinueSignal } from "../executer/signals.js";
import { Executable } from "./base.js";
import { LoopWithoutBodyError } from "../error/loop.js";
import { isTruthy } from "../executer/internal/isTruthy.js";
import { emitLoopIterationWarning, LOOP_WARNING_THRESHOLD } from "../util/loop-warning.js";
export class ConditionalLoop extends Executable {
  condition;
  body;
  tokens;
  static friendlyName = '반복 동안';
  constructor(condition, body, tokens){
    super(), this.condition = condition, this.body = body, this.tokens = tokens;
  }
  async execute(scope) {
    if (this.body.tokens.length === 0) {
      return;
    }
    let iterationCount = 0;
    let warned = false;
    try {
      while(true){
        const conditionValue = await this.condition.execute(scope);
        if (!isTruthy(conditionValue)) {
          break;
        }
        iterationCount += 1;
        if (!warned && iterationCount > LOOP_WARNING_THRESHOLD) {
          emitLoopIterationWarning({
            scope,
            tokens: this.tokens,
            iterations: iterationCount
          });
          warned = true;
        }
        await this.onRunChild({
          scope,
          childTokens: this.body.tokens,
          skipReport: true
        });
        try {
          await this.body.execute(scope);
        } catch (e) {
          if (e instanceof ContinueSignal) continue;
          throw e;
        }
      }
    } catch (e) {
      if (!(e instanceof BreakSignal)) {
        throw e;
      }
    }
  }
  validate(scope) {
    const childErrors = this.body.validate(scope);
    const conditionErrors = this.condition.validate(scope);
    const hasBodyError = this.body.children.length === 0 ? new LoopWithoutBodyError({
      tokens: this.tokens
    }) : null;
    return [
      ...conditionErrors,
      ...childErrors,
      hasBodyError
    ].filter(Boolean);
  }
}
