// @dalbit-yaksok/core/node/count-loop.js
import { LoopCountIsNotNumberError } from "../error/loop.js";
import { Scope } from "../executer/scope.js";
import { BreakSignal, ContinueSignal } from "../executer/signals.js";
import { NumberValue } from "../value/primitive.js";
import { Executable } from "./base.js";
import { emitLoopIterationWarning, LOOP_WARNING_THRESHOLD } from "../util/loop-warning.js";
export class CountLoop extends Executable {
  count;
  body;
  tokens;
  static friendlyName = '횟수 반복';
  constructor(count, body, tokens){
    super(), this.count = count, this.body = body, this.tokens = tokens;
  }
  async execute(_scope) {
    const scope = new Scope({
      parent: _scope,
      callerNode: this
    });
    const countValue = await this.count.execute(scope);
    if (!(countValue instanceof NumberValue)) {
      throw new LoopCountIsNotNumberError({
        tokens: this.count.tokens,
        value: countValue
      });
    }
    const countNumber = countValue.value;
    let warned = false;
    try {
      for(let i = 0; i < countNumber; i++){
        if (scope.codeFile?.session?.canRunNode) {
          if (!await scope.codeFile?.session?.canRunNode(scope, this.body)) {
            return;
          }
        }
        const iterationCount = i + 1;
        if (!warned && iterationCount > LOOP_WARNING_THRESHOLD) {
          emitLoopIterationWarning({
            scope,
            tokens: this.tokens,
            iterations: iterationCount
          });
          warned = true;
        }
        await this.onRunChild({
          scope,
          childTokens: this.body.tokens,
          skipReport: true
        });
        try {
          await this.body.execute(scope);
        } catch (e) {
          if (e instanceof ContinueSignal) continue;
          throw e;
        }
      }
    } catch (e) {
      if (!(e instanceof BreakSignal)) throw e;
    }
  }
  validate(scope) {
    return this.body.validate(scope);
  }
}
