// @dalbit-yaksok/core/node/count-loop.ts
import { YaksokError } from '../error/common.js'
import { LoopCountIsNotNumberError } from '../error/loop.js'
import { Scope } from '../executer/scope.js'
import { BreakSignal, ContinueSignal } from '../executer/signals.js'
import { Token } from '../prepare/tokenize/token.js'
import { NumberValue } from '../value/primitive.js'
import { Evaluable, Executable } from './base.js'
import { Block } from './block.js'
import {
    emitLoopIterationWarning,
    LOOP_WARNING_THRESHOLD,
} from '../util/loop-warning.js'

export class CountLoop extends Executable {
    static override friendlyName = '횟수 반복'

    constructor(
        public count: Evaluable,
        public body: Block,
        public override tokens: Token[],
    ) {
        super()
    }

    override async execute(_scope: Scope): Promise<void> {
        const scope = new Scope({
            parent: _scope,
            callerNode: this,
        })

        const countValue = await this.count.execute(scope)

        if (!(countValue instanceof NumberValue)) {
            throw new LoopCountIsNotNumberError({
                tokens: this.count.tokens,
                value: countValue,
            })
        }

        const countNumber = countValue.value

        let warned = false

        try {
            for (let i = 0; i < countNumber; i++) {
                if (scope.codeFile?.session?.canRunNode) {
                    if (
                        !(await scope.codeFile?.session?.canRunNode(
                            scope,
                            this.body,
                        ))
                    ) {
                        return
                    }
                }

                const iterationCount = i + 1

                if (!warned && iterationCount > LOOP_WARNING_THRESHOLD) {
                    emitLoopIterationWarning({
                        scope,
                        tokens: this.tokens,
                        iterations: iterationCount,
                    })
                    warned = true
                }

                await this.onRunChild({
                    scope,
                    childTokens: this.body.tokens,
                    skipReport: true,
                })
                try {
                    await this.body.execute(scope)
                } catch (e) {
                    if (e instanceof ContinueSignal) continue
                    throw e
                }
            }
        } catch (e) {
            if (!(e instanceof BreakSignal)) throw e
        }
    }

    override validate(scope: Scope): YaksokError[] {
        return this.body.validate(scope)
    }
}
