// @dalbit-yaksok/core/node/dict.js
import { IndexedValue } from "../value/indexed.js";
import { Evaluable, Expression } from "./base.js";
export class KeyValuePair extends Expression {
  key;
  entry;
  tokens;
  static friendlyName = '키-값 쌍';
  constructor(key, entry, tokens = []){
    super(String(key), tokens), this.key = key, this.entry = entry, this.tokens = tokens;
  }
}
export class KeyValuePairSequence extends Expression {
  pairs;
  tokens;
  static friendlyName = '키-값 쌍 목록';
  constructor(pairs, tokens = []){
    super('키-값 쌍 목록', tokens), this.pairs = pairs, this.tokens = tokens;
  }
}
export class DictLiteral extends Evaluable {
  pairs;
  tokens;
  static friendlyName = '사전';
  constructor(pairs, tokens = []){
    super(), this.pairs = pairs, this.tokens = tokens;
  }
  validate(scope) {
    const errors = this.pairs.flatMap((pair)=>pair.entry.validate(scope));
    return errors;
  }
  async execute(scope) {
    const evaluatedEntries = new Map(await Promise.all(this.pairs.map(async (pair)=>[
        pair.key,
        await pair.entry.execute(scope)
      ])));
    const value = new IndexedValue(evaluatedEntries);
    return value;
  }
}
