// @dalbit-yaksok/core/node/event.js
import { Scope, YaksokError } from "../mod.js";
import { errorToMachineReadable, renderErrorString } from "../error/render-error-string.js";
import { Executable } from "./base.js";
import { evaluateParams } from "./function.js";
export class DeclareEvent extends Executable {
  tokens;
  static friendlyName = '새 이벤트 만들기';
  eventId;
  name;
  constructor(props, tokens){
    super(), this.tokens = tokens;
    this.eventId = props.eventId;
    this.name = props.name;
  }
  execute(_scope) {
    return Promise.resolve();
  }
  validate(_scope) {
    return [];
  }
}
export class SubscribeEvent extends Executable {
  tokens;
  static friendlyName = '이벤트 구독하기';
  eventId;
  body;
  params;
  target;
  callerScope;
  constructor(props, tokens){
    super(), this.tokens = tokens;
    this.eventId = props.eventId;
    this.body = props.body;
    this.params = props.params;
    this.target = props.target;
  }
  async execute(scope) {
    const param = await evaluateParams(this.params, scope);
    if (this.target) {
      param['자신'] = await this.target.execute(scope);
    }
    const bodyParentScope = this.callerScope ?? scope;
    const session = scope.codeFile?.session;
    session?.aliveListeners.push(new Promise((resolve)=>{
      session?.eventCreation.pub(this.eventId, [
        param,
        ()=>{
          const subScope = new Scope({
            parent: bodyParentScope,
            callerNode: this,
            initialVariable: param
          });
          return this.body.execute(subScope).catch((e)=>{
            if (e instanceof YaksokError) {
              if (!e.codeFile) {
                e.codeFile = scope.codeFile;
              }
              session?.stderr(renderErrorString(e), errorToMachineReadable(e));
            }
            resolve();
          });
        },
        ()=>{
          resolve();
        },
        scope
      ]);
    }));
    return Promise.resolve();
  }
  validate(scope) {
    if (this.target) {
      return this.target.validate(scope);
    }
    return [];
  }
}
