// @dalbit-yaksok/core/node/ffi.js
import { YaksokError } from "../error/common.js";
import { FFIObject } from "../value/ffi.js";
import { Executable, Node } from "./base.js";
export class FFIBody extends Node {
  code;
  tokens;
  static friendlyName = '번역할 내용';
  constructor(code, tokens){
    super(), this.code = code, this.tokens = tokens;
  }
  validate() {
    return [];
  }
}
export class DeclareFFI extends Executable {
  tokens;
  static friendlyName = '번역 만들기';
  name;
  body;
  runtime;
  paramNames;
  dotReceiverTypeNames;
  constructor(props, tokens){
    super(), this.tokens = tokens;
    this.name = props.name;
    this.body = props.body;
    this.runtime = props.runtime;
    this.paramNames = props.paramNames || [];
    this.dotReceiverTypeNames = props.dotReceiverTypeNames;
    this.position = props.position;
  }
  execute(scope) {
    try {
      scope.addFunctionObject(this.toFFIObject(scope));
      return Promise.resolve();
    } catch (e) {
      if (e instanceof YaksokError && !e.tokens) {
        e.tokens = this.tokens;
      }
      throw e;
    }
  }
  toFFIObject(scope) {
    const codeFile = scope.codeFile;
    const ffiObject = new FFIObject(this.name, this.body, this.runtime, codeFile, {
      dotReceiverTypeNames: this.dotReceiverTypeNames
    });
    ffiObject.paramNames = this.paramNames;
    return ffiObject;
  }
  validate(scope) {
    try {
      const ffiObject = new FFIObject(this.name, 'VALIDATION', 'VALIDATION', undefined, {
        dotReceiverTypeNames: this.dotReceiverTypeNames
      });
      ffiObject.paramNames = this.paramNames;
      scope.addFunctionObject(ffiObject);
    } catch (error) {
      if (error instanceof YaksokError) {
        error.tokens = this.tokens;
        return [
          error
        ];
      } else {
        throw error;
      }
    }
    return [];
  }
}
