// @dalbit-yaksok/core/node/IfStatement.js
import { Executable } from "./base.js";
import { YaksokError } from "../error/common.js";
import { isTruthy } from "../executer/internal/isTruthy.js";
class StandaloneElseError extends YaksokError {
  constructor(props){
    super(props);
    this.message = `"아니면"은 "만약" 조건문의 본문이 있어야 사용할 수 있어요.`;
  }
}
class StandaloneElseIfError extends YaksokError {
  constructor(props){
    super(props);
    this.message = `"아니면 만약"은 "만약" 조건문의 본문이 있어야 사용할 수 있어요.`;
  }
}
export class IfStatement extends Executable {
  cases;
  tokens;
  static friendlyName = '조건문(만약)';
  constructor(cases, tokens){
    super(), this.cases = cases, this.tokens = tokens;
  }
  async execute(scope) {
    for (const { condition, body } of this.cases){
      if (scope.codeFile?.session?.canRunNode) {
        if (!await scope.codeFile?.session?.canRunNode(scope, this)) {
          return;
        }
      }
      const shouldStop = await this.shouldStop(condition, scope);
      if (!shouldStop) continue;
      await body.execute(scope);
      break;
    }
  }
  async shouldStop(condition, scope) {
    return !condition || isTruthy(await condition.execute(scope));
  }
  validate(scope) {
    const errors = this.cases.flatMap((caseItem)=>{
      const { condition, body } = caseItem;
      const conditionErrors = condition?.validate(scope) || [];
      const bodyErrors = body?.validate(scope) || [];
      return conditionErrors.concat(bodyErrors);
    });
    return errors;
  }
}
export class ElseStatement extends Executable {
  body;
  tokens;
  static friendlyName = '조건문(아니면)';
  constructor(body, tokens){
    super(), this.body = body, this.tokens = tokens;
  }
  validate(scope) {
    return [
      new StandaloneElseError({
        tokens: this.tokens
      }),
      ...this.body.validate(scope)
    ];
  }
}
export class ElseIfStatement extends Executable {
  elseIfCase;
  tokens;
  static friendlyName = '조건문(아니면 만약)';
  constructor(elseIfCase, tokens){
    super(), this.elseIfCase = elseIfCase, this.tokens = tokens;
  }
  validate(scope) {
    return [
      new StandaloneElseIfError({
        tokens: this.tokens
      }),
      ...this.elseIfCase.condition?.validate(scope) || [],
      ...this.elseIfCase.body.validate(scope)
    ];
  }
}
