// @dalbit-yaksok/core/node/lambda.ts
import { YaksokError } from '../error/common.js'
import { Scope } from '../executer/scope.js'
import type { Token } from '../prepare/tokenize/token.ts'
import { ValueType } from '../value/base.js'
import { LambdaFunctionValue } from '../value/lambda-function.js'
import { Evaluable } from './base.js'

export class LambdaLiteral extends Evaluable<LambdaFunctionValue> {
    static override friendlyName = '람다'

    constructor(
        public paramNames: string[],
        public body: Evaluable,
        public override tokens: Token[],
    ) {
        super()
    }

    override async execute(scope: Scope): Promise<LambdaFunctionValue> {
        return new LambdaFunctionValue(this.paramNames, this.body, scope)
    }

    override validate(scope: Scope): YaksokError[] {
        const lambdaScope = new Scope({
            parent: scope,
            initialVariable: Object.fromEntries(
                this.paramNames.map((name) => [name, new ValueType()]),
            ),
            callerNode: this,
        })

        return this.body.validate(lambdaScope)
    }
}
