// @dalbit-yaksok/core/node/list.js
import { ListIndexTypeError, StringIndexOutOfRangeError, TargetIsNotIndexedValueError } from "../error/indexed.js";
import { TupleNotMutableError } from "../error/indexed.js";
import { IndexedValue } from "../value/indexed.js";
import { ListValue } from "../value/list.js";
import { TupleValue } from "../value/tuple.js";
import { NumberValue, StringValue } from "../value/primitive.js";
import { Evaluable, Executable, Node } from "./base.js";
import { YaksokError } from "../error/common.js";
import { NotExecutableNodeError } from "../error/unknown-node.js";
import { assignerToOperatorMap } from "./operator.js";
export class Sequence extends Node {
  items;
  tokens;
  static friendlyName = '나열된 값';
  constructor(items, tokens){
    super(), this.items = items, this.tokens = tokens;
  }
  validate() {
    return [
      new NotExecutableNodeError({
        tokens: this.tokens,
        resource: {
          node: this
        }
      })
    ];
  }
}
export class ListLiteral extends Evaluable {
  items;
  tokens;
  static friendlyName = '목록';
  constructor(items, tokens){
    super(), this.items = items, this.tokens = tokens;
  }
  async execute(scope) {
    const evaluatedItems = await Promise.all(this.items.map((item)=>item.execute(scope)));
    const value = new ListValue(evaluatedItems);
    return value;
  }
  validate(scope) {
    const errors = this.items.flatMap((item)=>item.validate(scope)).filter((error)=>!!error);
    return errors;
  }
}
export class TupleLiteral extends Evaluable {
  items;
  tokens;
  static friendlyName = '튜플';
  constructor(items, tokens){
    super(), this.items = items, this.tokens = tokens;
  }
  async execute(scope) {
    const evaluatedItems = await Promise.all(this.items.map((item)=>item.execute(scope)));
    const value = new TupleValue(evaluatedItems);
    return value;
  }
  validate(scope) {
    const errors = this.items.flatMap((item)=>item.validate(scope)).filter((error)=>!!error);
    return errors;
  }
}
export class IndexFetch extends Evaluable {
  list;
  index;
  tokens;
  static friendlyName = '사전에서 값 가져오기';
  constructor(list, index, tokens){
    super(), this.list = list, this.index = index, this.tokens = tokens;
  }
  async execute(scope) {
    const list = await this.list.execute(scope);
    const index = await this.index.execute(scope);
    if (list instanceof StringValue) {
      if (!(index instanceof NumberValue)) {
        throw new ListIndexTypeError({
          tokens: this.index.tokens,
          resource: {
            index: index.toPrint()
          }
        });
      }
      try {
        ListValue.assertProperIndex(index.value);
      } catch (error) {
        if (error instanceof YaksokError && !error.tokens) {
          error.tokens = this.index.tokens;
        }
        throw error;
      }
      if (index.value >= list.value.length) {
        throw new StringIndexOutOfRangeError({
          tokens: this.tokens,
          resource: {
            target: list,
            index: index.value,
            length: list.value.length
          }
        });
      }
      return new StringValue(list.value[index.value]);
    }
    if (!(list instanceof IndexedValue)) {
      throw new TargetIsNotIndexedValueError({
        tokens: this.tokens,
        resource: {
          target: this.list
        }
      });
    }
    if (index instanceof ListValue) {
      const values = list.getItemsFromKeys(index);
      return values;
    }
    try {
      const value = list.getItem(index.value);
      return value;
    } catch (error) {
      if (error instanceof YaksokError && !error.tokens) {
        if (error instanceof ListIndexTypeError) {
          error.tokens = this.index.tokens;
        } else {
          error.tokens = this.tokens;
        }
      }
      throw error;
    }
  }
  async setValue(scope, value) {
    const list = await this.list.execute(scope);
    const index = await this.index.execute(scope);
    if (list instanceof TupleValue) {
      throw new TupleNotMutableError({
        tokens: this.tokens,
        resource: {}
      });
    }
    if (!(list instanceof IndexedValue)) {
      throw new TargetIsNotIndexedValueError({
        tokens: this.tokens.slice(0, -1),
        resource: {
          target: this.list
        }
      });
    }
    if (!(index instanceof NumberValue) && !(index instanceof StringValue)) {
      throw new ListIndexTypeError({
        tokens: this.tokens,
        resource: {
          index: index.toPrint()
        }
      });
    }
    list.setItem(index.value, value);
  }
  validate(scope) {
    const errors = [
      ...this.list.validate(scope) || [],
      ...this.index.validate(scope) || []
    ];
    return errors;
  }
}
export class SetToIndex extends Executable {
  target;
  value;
  operator;
  tokens;
  static friendlyName = '목록에 값 넣기';
  constructor(target, value, operator, tokens){
    super(), this.target = target, this.value = value, this.operator = operator, this.tokens = tokens;
    this.position = target.position;
  }
  async execute(scope) {
    const operatorNode = assignerToOperatorMap[this.operator];
    const operand = await this.value.execute(scope);
    let newValue = operand;
    if (operatorNode) {
      const oldValue = await this.target.execute(scope);
      const tempOperator = new operatorNode(this.tokens);
      try {
        newValue = await tempOperator.call(()=>Promise.resolve(oldValue), ()=>Promise.resolve(operand));
      } catch (error) {
        if (error instanceof YaksokError) {
          if (!error.tokens) {
            error.tokens = this.tokens;
          }
          if (!error.codeFile) {
            error.codeFile = scope.codeFile;
          }
        }
        throw error;
      }
    }
    await this.target.setValue(scope, newValue);
  }
  validate(scope) {
    const errors = [
      ...this.target.validate(scope) || [],
      ...this.value.validate(scope) || []
    ];
    return errors;
  }
}
