// @dalbit-yaksok/core/node/listLoop.js
import { NotEnumerableValueForListLoopError } from "../error/index.js";
import { BreakSignal, ContinueSignal } from "../executer/signals.js";
import { Executable } from "./base.js";
import { Scope } from "../executer/scope.js";
import { IndexedValue } from "../value/indexed.js";
import { NumberValue, StringValue } from "../value/primitive.js";
import { emitLoopIterationWarning, LOOP_WARNING_THRESHOLD } from "../util/loop-warning.js";
export class ListLoop extends Executable {
  list;
  variableName;
  body;
  tokens;
  static friendlyName = '목록 반복';
  constructor(list, variableName, body, tokens){
    super(), this.list = list, this.variableName = variableName, this.body = body, this.tokens = tokens;
  }
  async execute(_scope) {
    const scope = new Scope({
      parent: _scope,
      callerNode: this
    });
    const list = await this.list.execute(scope);
    this.assertRepeatTargetIsList(list);
    let iterationCount = 0;
    let warned = false;
    try {
      for (const value of list.enumerate()){
        if (scope.codeFile?.session?.canRunNode) {
          if (!await scope.codeFile?.session?.canRunNode(scope, this.body)) {
            return;
          }
        }
        iterationCount += 1;
        if (!warned && iterationCount > LOOP_WARNING_THRESHOLD) {
          emitLoopIterationWarning({
            scope,
            tokens: this.tokens,
            iterations: iterationCount
          });
          warned = true;
        }
        await this.onRunChild({
          scope,
          childTokens: this.body.tokens,
          skipReport: true
        });
        scope.setVariable(this.variableName, value, this.tokens);
        try {
          await this.body.execute(scope);
        } catch (e) {
          if (e instanceof ContinueSignal) continue;
          throw e;
        }
      }
    } catch (e) {
      if (!(e instanceof BreakSignal)) throw e;
    }
  }
  assertRepeatTargetIsList(target) {
    if (target instanceof IndexedValue || target instanceof StringValue) {
      return;
    }
    throw new NotEnumerableValueForListLoopError({
      resource: {
        value: target
      },
      tokens: this.list.tokens
    });
  }
  validate(scope) {
    const listScope = new Scope({
      parent: scope,
      callerNode: this,
      initialVariable: {
        [this.variableName]: new NumberValue(0)
      }
    });
    const listErrors = this.list.validate(scope);
    const bodyErrors = this.body.validate(listScope);
    return [
      ...listErrors,
      ...bodyErrors
    ];
  }
}
