// @dalbit-yaksok/core/node/loop.js
import { BreakSignal, ContinueSignal } from "../executer/signals.js";
import { Executable } from "./base.js";
import { LoopWithoutBodyError, NoBreakOrReturnError } from "../error/loop.js";
import { TOKEN_TYPE } from "../prepare/tokenize/token.js";
import { emitLoopIterationWarning, LOOP_WARNING_THRESHOLD } from "../util/loop-warning.js";
export class Loop extends Executable {
  body;
  tokens;
  static friendlyName = '반복';
  constructor(body, tokens){
    super(), this.body = body, this.tokens = tokens;
  }
  async execute(scope) {
    if (this.body.tokens.length === 0) {
      return;
    }
    let iterationCount = 0;
    let warned = false;
    try {
      while(true){
        if (scope.codeFile?.session?.canRunNode) {
          if (!await scope.codeFile?.session?.canRunNode(scope, this.body)) {
            return;
          }
        }
        iterationCount += 1;
        if (!warned && iterationCount > LOOP_WARNING_THRESHOLD) {
          emitLoopIterationWarning({
            scope,
            tokens: this.tokens,
            iterations: iterationCount
          });
          warned = true;
        }
        await this.onRunChild({
          scope,
          childTokens: this.body.tokens,
          skipReport: true
        });
        try {
          await this.body.execute(scope);
        } catch (e) {
          if (e instanceof ContinueSignal) continue;
          throw e;
        }
      }
    } catch (e) {
      if (!(e instanceof BreakSignal)) {
        throw e;
      }
    }
  }
  validate(scope) {
    const noBreakOrReturnError = hasBreakOrReturn(this, scope) ? [] : [
      new NoBreakOrReturnError({
        tokens: this.tokens
      })
    ];
    const childErrors = this.body.validate(scope);
    const hasBodyError = this.body.children.length === 0 ? new LoopWithoutBodyError({
      tokens: this.tokens
    }) : null;
    return [
      ...noBreakOrReturnError,
      ...childErrors,
      hasBodyError
    ].filter(Boolean);
  }
}
export class Break extends Executable {
  tokens;
  static friendlyName = '그만';
  constructor(tokens){
    super(), this.tokens = tokens;
  }
  execute(_scope) {
    throw new BreakSignal(this.tokens);
  }
  validate() {
    return [];
  }
}
export class Continue extends Executable {
  tokens;
  static friendlyName = '다음 반복';
  constructor(tokens){
    super(), this.tokens = tokens;
  }
  execute(_scope) {
    throw new ContinueSignal(this.tokens);
  }
  validate() {
    return [];
  }
}
function hasBreakOrReturn(node, scope) {
  const shouldSkipValidation = scope.codeFile?.session?.flags['skip-validate-break-or-return-in-loop'];
  if (shouldSkipValidation) {
    return true;
  }
  return node.tokens.some((token, index)=>{
    if (token.type === TOKEN_TYPE.IDENTIFIER && (token.value === '반복' || token.value === '약속')) {
      const nextToken = node.tokens[index + 1];
      if (nextToken && nextToken.type === TOKEN_TYPE.IDENTIFIER && nextToken.value === '그만') {
        return true;
      }
    }
    if (token.type === TOKEN_TYPE.IDENTIFIER && token.value === '반환하기') {
      return true;
    }
  });
}
