// @dalbit-yaksok/core/node/mention.js
import { YaksokError } from "../error/common.js";
import { ErrorInModuleError } from "../error/index.js";
import { Evaluable, Node } from "./base.js";
import { SubscribeEvent } from "./event.js";
import { evaluateParams, FunctionInvoke } from "./function.js";
import { IncompleteMentionError } from "../error/unknown-node.js";
export class Mention extends Node {
  value;
  tokens;
  static friendlyName = '불러올 파일 이름';
  constructor(value, tokens){
    super(), this.value = value, this.tokens = tokens;
  }
  toPrint() {
    return '@' + this.value;
  }
  validate(scope) {
    const error = new IncompleteMentionError({
      tokens: this.tokens,
      resource: {
        node: this
      }
    });
    error.codeFile = scope.codeFile;
    return [
      error
    ];
  }
}
export class MentionScope extends Evaluable {
  fileName;
  child;
  tokens;
  static friendlyName = '불러오기';
  constructor(fileName, child, tokens){
    super(), this.fileName = fileName, this.child = child, this.tokens = tokens;
  }
  async execute(scope) {
    const moduleCodeFile = scope.codeFile.session.getCodeFile(this.fileName);
    try {
      const moduleFileScope = await moduleCodeFile.run();
      if (this.child instanceof FunctionInvoke) {
        const evaluatedParams = await evaluateParams(this.child.params, scope);
        return await this.child.execute(moduleFileScope, evaluatedParams);
      }
      if (this.child instanceof SubscribeEvent) {
        this.child.callerScope = scope;
        await this.child.execute(moduleFileScope);
        return undefined;
      }
      return await this.child.execute(moduleFileScope);
    } catch (error) {
      if (error instanceof YaksokError) {
        error.codeFile = moduleCodeFile;
        throw new ErrorInModuleError({
          resource: {
            fileName: this.fileName
          },
          tokens: this.tokens,
          child: error
        });
      }
      throw error;
    }
  }
  toPrint() {
    return '@' + this.fileName + ' ' + this.child.toPrint();
  }
  validate(scope) {
    const moduleCodeFile = scope.codeFile.session.getCodeFile(this.fileName);
    let validatingScope;
    let moduleErrors = [];
    try {
      const validationResult = moduleCodeFile.validate();
      validatingScope = validationResult.validatingScope;
      moduleErrors = validationResult.errors;
    } catch (error) {
      if (error instanceof YaksokError) {
        error.codeFile = moduleCodeFile;
        const errorInstance = new ErrorInModuleError({
          resource: {
            fileName: this.fileName
          },
          tokens: this.tokens,
          child: error
        });
        errorInstance.codeFile = scope.codeFile;
        return [
          errorInstance
        ];
      }
      throw error;
    }
    let childErrors = [];
    if (this.child instanceof FunctionInvoke) {
      for(const paramName in this.child.params){
        const param = this.child.params[paramName];
        childErrors = childErrors.concat(param.validate(scope));
      }
    } else {
      // validatingScope가 undefined가 아닐 때만 자식 노드의 validate 호출
      if (validatingScope) {
        childErrors = childErrors.concat(this.child.validate(validatingScope));
      }
    }
    return [
      ...moduleErrors,
      ...childErrors
    ];
  }
}
