// @dalbit-yaksok/core/node/misc.js
import { Evaluable, Executable, Node } from "./base.js";
import { StringValue } from "../value/primitive.js";
export class EOL extends Node {
  tokens;
  static friendlyName = '줄바꿈';
  constructor(tokens){
    super(), this.tokens = tokens;
  }
  validate() {
    return [];
  }
}
export class Indent extends Node {
  size;
  tokens;
  static friendlyName = '들여쓰기';
  constructor(size, tokens){
    super(), this.size = size, this.tokens = tokens;
  }
  validate() {
    return [];
  }
}
export class Print extends Executable {
  value;
  tokens;
  static friendlyName = '보여주기';
  constructor(value, tokens){
    super(), this.value = value, this.tokens = tokens;
  }
  async execute(scope) {
    const printFunction = scope.codeFile?.session?.stdout ?? console.log;
    const evaluated = await this.value.execute(scope);
    printFunction(evaluated.toPrint());
  }
  validate(scope) {
    return this.value.validate(scope);
  }
}
export class TypeOf extends Evaluable {
  value;
  tokens;
  static friendlyName = '값 종류';
  constructor(value, tokens){
    super(), this.value = value, this.tokens = tokens;
  }
  async execute(scope) {
    const evaluated = await this.value.execute(scope);
    return new StringValue(evaluated.constructor.friendlyName);
  }
  validate(scope) {
    return this.value.validate(scope);
  }
}
export class Pause extends Executable {
  tokens;
  static friendlyName = '잠깐 멈추기';
  constructor(tokens){
    super(), this.tokens = tokens;
  }
  execute(scope) {
    scope.codeFile?.session?.pubsub.pub('debug', [
      scope,
      this
    ]);
    scope.codeFile?.session?.pause();
    return Promise.resolve();
  }
  validate(_scope) {
    return [];
  }
}
