// @dalbit-yaksok/core/node/operator.js
import { InvalidTypeForCompareError } from "../error/calculation.js";
import { InvalidTypeForOperatorError, RangeEndMustBeIntegerError, RangeEndMustBeNumberError, RangeStartMustBeIntegerError, RangeStartMustBeLessThanEndError, RangeStartMustBeNumberError } from "../error/index.js";
import { PrimitiveValue } from "../value/base.js";
import { ListValue } from "../value/list.js";
import { BooleanValue, NumberValue, StringValue } from "../value/primitive.js";
import { Operator } from "./base.js";
import { cleanFloatingPointError } from "../util/float-precision.js";
export class PlusOperator extends Operator {
  tokens;
  static friendlyName = '더하기(+)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '+';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new NumberValue(cleanFloatingPointError(leftValue.value + rightValue.value));
    }
    if (leftValue instanceof StringValue && rightValue instanceof StringValue) {
      return new StringValue(leftValue.value + rightValue.value);
    }
    if (leftValue instanceof StringValue && rightValue instanceof NumberValue) {
      return new StringValue(leftValue.value + rightValue.value.toString());
    }
    if (leftValue instanceof NumberValue && rightValue instanceof StringValue) {
      return new StringValue(leftValue.value.toString() + rightValue.value);
    }
    if (leftValue instanceof ListValue && rightValue instanceof ListValue) {
      return new ListValue([
        ...Array.from(leftValue.enumerate()),
        ...Array.from(rightValue.enumerate())
      ]);
    }
    throw new InvalidTypeForOperatorError({
      position: this.tokens?.[0].position,
      resource: {
        operator: this,
        operands: [
          leftValue,
          rightValue
        ]
      }
    });
  }
}
export class MinusOperator extends Operator {
  tokens;
  static friendlyName = '빼기(-)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '-';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new NumberValue(cleanFloatingPointError(leftValue.value - rightValue.value));
    }
    throw new InvalidTypeForOperatorError({
      position: this.tokens?.[0].position,
      resource: {
        operator: this,
        operands: [
          leftValue,
          rightValue
        ]
      }
    });
  }
}
export class MultiplyOperator extends Operator {
  tokens;
  static friendlyName = '곱하기(*)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '*';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    const ensureValidListMultiplier = (value)=>{
      if (!Number.isSafeInteger(value) || value < 0) {
        throw new InvalidTypeForOperatorError({
          position: this.tokens?.[0].position,
          resource: {
            operator: this,
            operands: [
              leftValue,
              rightValue
            ]
          }
        });
      }
      return value;
    };
    const repeatList = (list, multiplier)=>{
      const sourceElements = Array.from(list.enumerate());
      const repeatedElements = [];
      for(let i = 0; i < multiplier; i++){
        for (const element of sourceElements){
          repeatedElements.push(element);
        }
      }
      return new ListValue(repeatedElements);
    };
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new NumberValue(cleanFloatingPointError(leftValue.value * rightValue.value));
    }
    if (leftValue instanceof StringValue && rightValue instanceof NumberValue) {
      return new StringValue(leftValue.value.repeat(rightValue.value));
    }
    if (leftValue instanceof NumberValue && rightValue instanceof StringValue) {
      return new StringValue(rightValue.value.repeat(leftValue.value));
    }
    if (leftValue instanceof ListValue && rightValue instanceof NumberValue) {
      const multiplier = ensureValidListMultiplier(rightValue.value);
      return repeatList(leftValue, multiplier);
    }
    if (leftValue instanceof NumberValue && rightValue instanceof ListValue) {
      const multiplier = ensureValidListMultiplier(leftValue.value);
      return repeatList(rightValue, multiplier);
    }
    throw new InvalidTypeForOperatorError({
      position: this.tokens?.[0].position,
      resource: {
        operator: this,
        operands: [
          leftValue,
          rightValue
        ]
      }
    });
  }
}
export class DivideOperator extends Operator {
  tokens;
  static friendlyName = '나누기(/)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '/';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new NumberValue(cleanFloatingPointError(leftValue.value / rightValue.value));
    }
    throw new InvalidTypeForOperatorError({
      position: this.tokens?.[0].position,
      resource: {
        operator: this,
        operands: [
          leftValue,
          rightValue
        ]
      }
    });
  }
}
export class ModularOperator extends Operator {
  tokens;
  static friendlyName = '나머지(%)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '%';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new NumberValue(cleanFloatingPointError(leftValue.value % rightValue.value));
    }
    throw new InvalidTypeForOperatorError({
      position: this.tokens?.[0].position,
      resource: {
        operator: this,
        operands: [
          leftValue,
          rightValue
        ]
      }
    });
  }
}
export class PowerOperator extends Operator {
  tokens;
  static friendlyName = '제곱(**)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '**';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new NumberValue(cleanFloatingPointError(leftValue.value ** rightValue.value));
    }
    throw new InvalidTypeForOperatorError({
      position: this.tokens?.[0].position,
      resource: {
        operator: this,
        operands: [
          leftValue,
          rightValue
        ]
      }
    });
  }
}
export class IntegerDivideOperator extends Operator {
  tokens;
  static friendlyName = '정수 나누기(//)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '//';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new NumberValue(Math.floor(leftValue.value / rightValue.value));
    }
    throw new InvalidTypeForOperatorError({
      position: this.tokens?.[0].position,
      resource: {
        operator: this,
        operands: [
          leftValue,
          rightValue
        ]
      }
    });
  }
}
export class EqualOperator extends Operator {
  tokens;
  static friendlyName = '같다(==)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '==';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    const isSameType = leftValue.constructor === rightValue.constructor;
    const isBothPrimitive = leftValue instanceof PrimitiveValue && rightValue instanceof PrimitiveValue;
    if (!isSameType || !isBothPrimitive) {
      throw new InvalidTypeForCompareError({
        resource: {
          left: leftValue,
          right: rightValue
        },
        position: this.tokens?.[0].position
      });
    }
    return new BooleanValue(leftValue.value === rightValue.value);
  }
}
export class NotEqualOperator extends EqualOperator {
  static friendlyName = '같지 않다(!=)';
  toPrint() {
    return '!=';
  }
  async call(left, right) {
    return new BooleanValue(!(await super.call(left, right)).value);
  }
}
export class AndOperator extends Operator {
  tokens;
  static friendlyName = '이고';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '이고';
  }
  async call(left, right) {
    const leftValue = await left();
    if (!(leftValue instanceof BooleanValue)) {
      throw new InvalidTypeForOperatorError({
        position: this.tokens?.[0].position,
        resource: {
          operator: this,
          operands: [
            leftValue
          ]
        }
      });
    }
    if (!leftValue.value) return new BooleanValue(false);
    const rightValue = await right();
    if (!(rightValue instanceof BooleanValue)) {
      throw new InvalidTypeForOperatorError({
        position: this.tokens?.[0].position,
        resource: {
          operator: this,
          operands: [
            leftValue,
            rightValue
          ]
        }
      });
    }
    return new BooleanValue(rightValue.value);
  }
}
export class OrOperator extends Operator {
  tokens;
  static friendlyName = '이거나(거나)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '이거나(거나)';
  }
  async call(left, right) {
    const leftValue = await left();
    if (!(leftValue instanceof BooleanValue)) {
      throw new InvalidTypeForOperatorError({
        position: this.tokens?.[0].position,
        resource: {
          operator: this,
          operands: [
            leftValue
          ]
        }
      });
    }
    if (leftValue.value) return new BooleanValue(true);
    const rightValue = await right();
    if (!(rightValue instanceof BooleanValue)) {
      throw new InvalidTypeForOperatorError({
        position: this.tokens?.[0].position,
        resource: {
          operator: this,
          operands: [
            leftValue,
            rightValue
          ]
        }
      });
    }
    return new BooleanValue(rightValue.value);
  }
}
export class GreaterThanOperator extends Operator {
  tokens;
  static friendlyName = '크다(>)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '>';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new BooleanValue(leftValue.value > rightValue.value);
    }
    throw new InvalidTypeForCompareError({
      position: this.tokens?.[0].position,
      resource: {
        left: leftValue,
        right: rightValue
      }
    });
  }
}
export class LessThanOperator extends Operator {
  tokens;
  static friendlyName = '작다(<)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '<';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new BooleanValue(leftValue.value < rightValue.value);
    }
    throw new InvalidTypeForCompareError({
      position: this.tokens?.[0].position,
      resource: {
        left: leftValue,
        right: rightValue
      }
    });
  }
}
export class GreaterThanOrEqualOperator extends Operator {
  tokens;
  static friendlyName = '크거나 같다(>=)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '>=';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new BooleanValue(leftValue.value >= rightValue.value);
    }
    throw new InvalidTypeForCompareError({
      position: this.tokens?.[0].position,
      resource: {
        left: leftValue,
        right: rightValue
      }
    });
  }
}
export class LessThanOrEqualOperator extends Operator {
  tokens;
  static friendlyName = '작거나 같다(<=)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '<=';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    if (leftValue instanceof NumberValue && rightValue instanceof NumberValue) {
      return new BooleanValue(leftValue.value <= rightValue.value);
    }
    throw new InvalidTypeForCompareError({
      position: this.tokens?.[0].position,
      resource: {
        left: leftValue,
        right: rightValue
      }
    });
  }
}
export class RangeOperator extends Operator {
  tokens;
  static friendlyName = '범위에서 목록 만들기(~)';
  constructor(tokens){
    super(null, tokens), this.tokens = tokens;
  }
  toPrint() {
    return '~';
  }
  async call(left, right) {
    const leftValue = await left();
    const rightValue = await right();
    this.assertProperOperands([
      leftValue,
      rightValue
    ]);
    const [start, end] = [
      leftValue,
      rightValue
    ];
    const roundedStart = Math.round(start.value);
    const roundedEnd = Math.round(end.value);
    const items = new Array(roundedEnd - roundedStart + 1).fill(null).map((_, index)=>new NumberValue(roundedStart + index));
    return new ListValue(items);
  }
  assertProperOperands(operands) {
    const [start, end] = operands;
    this.assertProperStartType(start);
    this.assertProperEndType(end);
    this.assertRangeStartLessThanEnd(start.value, end.value);
  }
  assertProperStartType(start) {
    if (!(start instanceof NumberValue)) {
      throw new RangeStartMustBeNumberError({
        resource: {
          start
        }
      });
    }
    if (!Number.isInteger(start.value)) {
      throw new RangeStartMustBeIntegerError({
        resource: {
          start
        }
      });
    }
  }
  assertProperEndType(end) {
    if (!(end instanceof NumberValue)) {
      throw new RangeEndMustBeNumberError({
        resource: {
          end
        }
      });
    }
    if (!Number.isInteger(end.value)) {
      throw new RangeEndMustBeIntegerError({
        resource: {
          end
        }
      });
    }
  }
  assertRangeStartLessThanEnd(start, end) {
    if (start <= end) return;
    throw new RangeStartMustBeLessThanEndError({
      position: this.tokens[0].position,
      resource: {
        start,
        end
      }
    });
  }
}
export const assignerToOperatorMap = {
  '+=': PlusOperator,
  '-=': MinusOperator,
  '*=': MultiplyOperator,
  '/=': DivideOperator,
  '%=': ModularOperator
};
