// @dalbit-yaksok/core/node/primitive-literal.js
import { BooleanValue, NumberValue, StringValue } from "../value/primitive.js";
import { Evaluable, Node } from "./base.js";
export class NumberLiteral extends Evaluable {
  content;
  tokens;
  static friendlyName = '숫자';
  constructor(content, tokens){
    super(), this.content = content, this.tokens = tokens;
  }
  execute(_scope) {
    return Promise.resolve(new NumberValue(this.content));
  }
  toPrint() {
    return this.content.toString();
  }
  toNumber() {
    return this.content;
  }
  validate() {
    return [];
  }
}
export class StringLiteral extends Evaluable {
  content;
  tokens;
  static friendlyName = '문자';
  constructor(content, tokens){
    super(), this.content = content, this.tokens = tokens;
  }
  execute(_scope) {
    return Promise.resolve(new StringValue(this.content));
  }
  toPrint() {
    return this.content;
  }
  validate() {
    return [];
  }
}
export class BooleanLiteral extends Evaluable {
  content;
  tokens;
  static friendlyName = '참거짓';
  constructor(content, tokens){
    super(), this.content = content, this.tokens = tokens;
  }
  execute(_scope) {
    return Promise.resolve(new BooleanValue(this.content));
  }
  toPrint() {
    return this.content ? '참' : '거짓';
  }
  validate() {
    return [];
  }
}
export class TemplateStringPart extends Node {
  content;
  tokens;
  static friendlyName = '템플릿 문자열 부분';
  constructor(content, tokens){
    super(), this.content = content, this.tokens = tokens;
  }
  toPrint() {
    return this.content;
  }
  validate() {
    return [];
  }
}
export class TemplateLiteral extends Evaluable {
  parts;
  tokens;
  static friendlyName = '템플릿 문자열';
  constructor(parts, tokens){
    super(), this.parts = parts, this.tokens = tokens;
  }
  async execute(scope) {
    const results = [];
    for (const part of this.parts){
      if (part instanceof TemplateStringPart) {
        results.push(part.content);
      } else {
        const value = await part.execute(scope);
        results.push(value.toPrint());
      }
    }
    return new StringValue(results.join(''));
  }
  toPrint() {
    return this.parts.map((p)=>p.toPrint()).join('');
  }
  validate(scope) {
    return this.parts.filter((part)=>part instanceof Evaluable).flatMap((part)=>part.validate(scope));
  }
}
