// @dalbit-yaksok/core/node/typecast.js
import { InvalidTypeCastError } from "../error/typecast.js";
import { Evaluable } from "./base.js";
import { BooleanValue, NumberValue, StringValue } from "../value/primitive.js";
const TRUTHY_STRINGS = new Set([
  '참',
  'true',
  '맞음'
]);
const FALSY_STRINGS = new Set([
  '거짓',
  'false',
  '아님'
]);
export class TypeCast extends Evaluable {
  value;
  targetType;
  tokens;
  static friendlyName = '형변환';
  constructor(value, targetType, tokens){
    super(), this.value = value, this.targetType = targetType, this.tokens = tokens;
  }
  async execute(scope) {
    const evaluated = await this.value.execute(scope);
    switch(this.targetType){
      case '숫자':
        return this.castToNumber(evaluated);
      case '문자열':
        return this.castToString(evaluated);
      case '참거짓':
        return this.castToBoolean(evaluated);
    }
  }
  castToNumber(value) {
    if (value instanceof NumberValue) {
      return value;
    }
    if (value instanceof StringValue) {
      const parsed = parseFloat(value.value);
      if (isNaN(parsed)) {
        throw new InvalidTypeCastError({
          tokens: this.tokens,
          resource: {
            value,
            targetType: this.targetType
          }
        });
      }
      return new NumberValue(parsed);
    }
    if (value instanceof BooleanValue) {
      return new NumberValue(value.value ? 1 : 0);
    }
    throw new InvalidTypeCastError({
      tokens: this.tokens,
      resource: {
        value,
        targetType: this.targetType
      }
    });
  }
  castToString(value) {
    return new StringValue(value.toPrint());
  }
  castToBoolean(value) {
    if (value instanceof BooleanValue) {
      return value;
    }
    if (value instanceof StringValue) {
      const lower = value.value.toLowerCase();
      if (TRUTHY_STRINGS.has(lower)) return new BooleanValue(true);
      if (FALSY_STRINGS.has(lower)) return new BooleanValue(false);
      return new BooleanValue(value.value.length > 0);
    }
    if (value instanceof NumberValue) {
      return new BooleanValue(value.value !== 0);
    }
    return new BooleanValue(true);
  }
  validate(scope) {
    return this.value.validate(scope);
  }
}
