// @dalbit-yaksok/core/node/variable.js
import { YaksokError } from "../error/common.js";
import { Evaluable } from "./base.js";
import { NumberValue } from "../value/primitive.js";
import { assignerToOperatorMap } from "./operator.js";
import { assertValidIdentifierName } from "../util/assert-valid-identifier-name.js";
import { ClassValue, createValidationInstanceFromClass, NewInstance } from "./class.js";
export class SetVariable extends Evaluable {
  name;
  value;
  tokens;
  operator;
  static friendlyName = '변수 정하기';
  __kind;
  constructor(name, value, tokens, operator){
    super(), this.name = name, this.value = value, this.tokens = tokens, this.operator = operator, this.__kind = 'SetVariable';
    assertValidIdentifierName(name, tokens[0]);
  }
  async execute(scope) {
    const { name, value } = this;
    const operatorNode = assignerToOperatorMap[this.operator];
    const operand = await value.execute(scope);
    let newValue = operand;
    if (operatorNode) {
      const oldValue = scope.getVariable(name, this.tokens);
      const tempOperator = new operatorNode(this.tokens);
      try {
        newValue = await tempOperator.call(()=>Promise.resolve(oldValue), ()=>Promise.resolve(operand));
      } catch (error) {
        if (error instanceof YaksokError) {
          if (!error.tokens) {
            error.tokens = this.tokens;
          }
          if (!error.codeFile) {
            error.codeFile = scope.codeFile;
          }
        }
        throw error;
      }
    }
    scope.setVariable(name, newValue, this.tokens);
    return newValue;
  }
  validate(scope) {
    const errors = this.value.validate(scope);
    scope.setLocalVariable(this.name, inferValidationValue(this.value, scope));
    return errors;
  }
}
function inferValidationValue(value, scope) {
  if (value instanceof NewInstance) {
    try {
      const classValue = scope.getVariable(value.className, value.tokens);
      if (classValue instanceof ClassValue) {
        return createValidationInstanceFromClass(classValue, value.arguments_.length);
      }
    } catch  {
    // Ignore lookup errors and fall back to unknown numeric placeholder.
    }
  }
  return new NumberValue(0);
}
