// @dalbit-yaksok/core/prepare/tokenize/token.js
export var TOKEN_TYPE = /*#__PURE__*/ function(TOKEN_TYPE) {
  TOKEN_TYPE["NUMBER"] = "NUMBER";
  TOKEN_TYPE["ASSIGNER"] = "ASSIGNER";
  TOKEN_TYPE["STRING"] = "STRING";
  TOKEN_TYPE["OPERATOR"] = "OPERATOR";
  TOKEN_TYPE["SPACE"] = "SPACE";
  TOKEN_TYPE["INDENT"] = "INDENT";
  TOKEN_TYPE["IDENTIFIER"] = "IDENTIFIER";
  TOKEN_TYPE["COMMA"] = "COMMA";
  TOKEN_TYPE["COLON"] = "COLON";
  TOKEN_TYPE["OPENING_PARENTHESIS"] = "OPENING_PARENTHESIS";
  TOKEN_TYPE["CLOSING_PARENTHESIS"] = "CLOSING_PARENTHESIS";
  TOKEN_TYPE["OPENING_BRACKET"] = "OPENING_BRACKET";
  TOKEN_TYPE["CLOSING_BRACKET"] = "CLOSING_BRACKET";
  TOKEN_TYPE["OPENING_BRACE"] = "OPENING_BRACE";
  TOKEN_TYPE["CLOSING_BRACE"] = "CLOSING_BRACE";
  TOKEN_TYPE["FFI_BODY"] = "FFI_BODY";
  TOKEN_TYPE["NEW_LINE"] = "NEW_LINE";
  TOKEN_TYPE["LINE_COMMENT"] = "LINE_COMMENT";
  TOKEN_TYPE["MENTION"] = "MENTION";
  TOKEN_TYPE["DOT"] = "DOT";
  TOKEN_TYPE["UNKNOWN"] = "UNKNOWN";
  TOKEN_TYPE["REGEX"] = "REGEX";
  TOKEN_TYPE["TEMPLATE_STRING_START"] = "TEMPLATE_STRING_START";
  TOKEN_TYPE["TEMPLATE_STRING_PART"] = "TEMPLATE_STRING_PART";
  TOKEN_TYPE["TEMPLATE_STRING_END"] = "TEMPLATE_STRING_END";
  return TOKEN_TYPE;
}({});
export const TOKEN_TYPE_TO_TEXT = {
  [TOKEN_TYPE.NUMBER]: '숫자',
  [TOKEN_TYPE.STRING]: '문자',
  [TOKEN_TYPE.ASSIGNER]: '값 넣기',
  [TOKEN_TYPE.OPERATOR]: '연산자',
  [TOKEN_TYPE.SPACE]: '공백',
  [TOKEN_TYPE.INDENT]: '들여쓰기',
  [TOKEN_TYPE.IDENTIFIER]: '식별자',
  [TOKEN_TYPE.COMMA]: '쉼표',
  [TOKEN_TYPE.COLON]: '쌍점',
  [TOKEN_TYPE.OPENING_PARENTHESIS]: '여는 괄호',
  [TOKEN_TYPE.CLOSING_PARENTHESIS]: '닫는 괄호',
  [TOKEN_TYPE.OPENING_BRACKET]: '여는 대괄호',
  [TOKEN_TYPE.CLOSING_BRACKET]: '닫는 대괄호',
  [TOKEN_TYPE.OPENING_BRACE]: '여는 중괄호',
  [TOKEN_TYPE.CLOSING_BRACE]: '닫는 중괄호',
  [TOKEN_TYPE.FFI_BODY]: '번역 코드',
  [TOKEN_TYPE.NEW_LINE]: '줄바꿈',
  [TOKEN_TYPE.LINE_COMMENT]: '주석',
  [TOKEN_TYPE.MENTION]: '불러오기',
  [TOKEN_TYPE.DOT]: '온점',
  [TOKEN_TYPE.REGEX]: '정규식',
  [TOKEN_TYPE.UNKNOWN]: '알 수 없음',
  [TOKEN_TYPE.TEMPLATE_STRING_START]: '템플릿 문자열 시작',
  [TOKEN_TYPE.TEMPLATE_STRING_PART]: '템플릿 문자열 부분',
  [TOKEN_TYPE.TEMPLATE_STRING_END]: '템플릿 문자열 끝'
};
