// @dalbit-yaksok/core/value/function.js
import { ObjectValue } from "./base.js";
import { NumberValue } from "./primitive.js";
import { ReturnSignal } from "../executer/signals.js";
import { Scope } from "../executer/scope.js";
import { YaksokError } from "../error/common.js";
import { CallStackDepthExceededError } from "../error/function.js";
import { FEATURE_FLAG } from "../constant/feature-flags.js";
const DEFAULT_RETURN_VALUE = new NumberValue(0);
const MAX_CALL_STACK_DEPTH = 32;
export class FunctionObject extends ObjectValue {
  name;
  body;
  declaredScope;
  paramNames;
  options;
  static friendlyName = '약속';
  constructor(name, body, declaredScope, paramNames = [], options = {}){
    super(), this.name = name, this.body = body, this.declaredScope = declaredScope, this.paramNames = paramNames, this.options = options;
  }
  async run(args, callSiteScope) {
    const lexicalScope = this.declaredScope ?? callSiteScope;
    const previousDepth = callSiteScope?.callStackDepth ?? lexicalScope?.callStackDepth ?? 0;
    const nextDepth = previousDepth + 1;
    if (nextDepth > MAX_CALL_STACK_DEPTH) {
      // FEATURE FLAG 확인: DISABLE_CALL_STACK_DEPTH_LIMIT가 활성화되면 에러를 발생시키지 않음
      const codeFile = this.declaredScope?.codeFile ?? callSiteScope?.codeFile ?? lexicalScope?.codeFile;
      const isDisabled = codeFile?.session?.flags[FEATURE_FLAG.DISABLE_CALL_STACK_DEPTH_LIMIT] === true;
      if (!isDisabled) {
        const errorInstance = new CallStackDepthExceededError({
          resource: {
            limit: MAX_CALL_STACK_DEPTH,
            depth: nextDepth
          }
        });
        errorInstance.codeFile = codeFile;
        throw errorInstance;
      }
    }
    const functionScope = new Scope({
      parent: lexicalScope,
      initialVariable: args,
      callStackDepth: nextDepth,
      callerNode: this.body
    });
    try {
      await this.body.execute(functionScope);
    } catch (e) {
      if (e instanceof ReturnSignal) {
        return e.value || DEFAULT_RETURN_VALUE;
      }
      if (e instanceof YaksokError && !e.codeFile) {
        e.codeFile = this.declaredScope?.codeFile ?? callSiteScope?.codeFile ?? lexicalScope?.codeFile;
      }
      throw e;
    }
    return DEFAULT_RETURN_VALUE;
  }
}
