// @dalbit-yaksok/core/value/indexed.js
import { IndexKeyNotFoundError, ListIndexTypeError } from "../error/indexed.js";
import { ObjectValue } from "./base.js";
import { NumberValue, StringValue } from "./primitive.js";
export class IndexedValue extends ObjectValue {
  entries;
  static friendlyName = '사전';
  constructor(entries){
    super(), this.entries = entries;
  }
  getItem(index) {
    if (!this.entries.has(index)) {
      throw new IndexKeyNotFoundError({
        resource: {
          target: this,
          index
        }
      });
    }
    return this.entries.get(index);
  }
  setItem(index, value) {
    this.entries.set(index, value);
  }
  getEntries() {
    return this.entries.entries();
  }
  getItemsFromKeys(keysListValue) {
    const keys = [
      ...keysListValue.entries.values()
    ];
    const entries = new Map();
    for (const keyValue of keys){
      if (!(keyValue instanceof StringValue || keyValue instanceof NumberValue)) {
        throw new ListIndexTypeError({
          resource: {
            index: keyValue.toPrint()
          }
        });
      }
      const key = keyValue.value;
      const value = this.getItem(key);
      entries.set(key, value);
    }
    return new IndexedValue(entries);
  }
  toPrint() {
    return `{\n\t${[
      ...this.entries.entries()
    ].map(([key, value])=>{
      let valueString = value.toPrint();
      if (valueString.includes('\n')) {
        const lines = valueString.split('\n');
        valueString = [
          lines[0],
          ...lines.slice(1).map((l)=>'\t' + l)
        ].join('\n');
      }
      return `${key}: ${valueString}`;
    }).join('\n\t')}
}`;
  }
  enumerate() {
    const keys = this.entries.keys().toArray();
    const keysInStringValue = keys.map((k)=>typeof k === 'string' ? new StringValue(k) : new NumberValue(k));
    return keysInStringValue;
  }
}
