// @dalbit-yaksok/core/value/tuple.js
import { ListIndexMustBeGreaterOrEqualThan0Error, ListIndexTypeError } from "../error/indexed.js";
import { TupleNotMutableError } from "../error/indexed.js";
import { IndexedValue } from "./indexed.js";
import { NumberValue } from "./primitive.js";
export class TupleValue extends IndexedValue {
  static friendlyName = '튜플';
  entries = new Map();
  constructor(entries){
    const entriesMap = new Map(entries.map((entry, index)=>[
        index,
        entry
      ]));
    super(entriesMap);
    this.entries = entriesMap;
  }
  getItem(index) {
    TupleValue.assertProperIndex(index);
    return super.getItem(index);
  }
  setItem(_index, _value) {
    throw new TupleNotMutableError({
      resource: {}
    });
  }
  static assertProperIndex(index) {
    if (!Number.isSafeInteger(index)) {
      throw new ListIndexTypeError({
        resource: {
          index
        }
      });
    }
    const isProperIndex = index >= 0;
    if (isProperIndex) {
      return;
    }
    throw new ListIndexMustBeGreaterOrEqualThan0Error({
      resource: {
        index
      }
    });
  }
  toPrint() {
    const keys = this.entries.keys().toArray();
    if (keys.length === 0) {
      return '(빈 튜플)';
    }
    const maxKey = Math.max(...keys);
    const values = [];
    for(let i = 0; i <= maxKey; i++){
      values[i] = this.entries.get(i) ?? null;
    }
    const formatted = values.map((value)=>value ? value.toPrint() : '(값 없음)').join(', ');
    const trailingComma = keys.length === 1 ? ',' : '';
    return `(${formatted}${trailingComma})`;
  }
  enumerate() {
    return this.entries.values();
  }
  getItemsFromKeys(keysListValue) {
    const keys = [
      ...keysListValue.entries.values()
    ];
    const list = [];
    for (const keyValue of keys){
      if (!(keyValue instanceof NumberValue)) {
        throw new ListIndexTypeError({
          resource: {
            index: keyValue.toPrint()
          }
        });
      }
      const key = keyValue.value;
      const value = this.getItem(key);
      list.push(value);
    }
    return new TupleValue(list);
  }
}
