// @dalbit-yaksok/core/prepare/parse/rule/index.js
import { Formula, NotExpression, ValueWithParenthesis } from "../../../node/calculation.js";
import { TupleLiteral } from "../../../node/list.js";
import { AndOperator, Block, BooleanLiteral, Break, Continue, ConditionalLoop, DivideOperator, ElseIfStatement, ElseStatement, EOL, EqualOperator, Evaluable, Expression, FetchMember, FunctionInvoke, GreaterThanOperator, GreaterThanOrEqualOperator, Identifier, IfStatement, IndexFetch, IntegerDivideOperator, LambdaLiteral, LessThanOperator, LessThanOrEqualOperator, ListLiteral, Loop, MemberFunctionInvoke, MinusOperator, ModularOperator, MultiplyOperator, NewInstance, Operator, OrOperator, Pause, PlusOperator, PowerOperator, Print, RangeOperator, Sequence, SetMember, SetToIndex, SetVariable, TypeCast, TypeOf } from "../../../node/index.js";
import { NotEqualOperator } from "../../../node/operator.js";
import { TemplateLiteral, TemplateStringPart } from "../../../node/primitive-literal.js";
import { ReturnStatement } from "../../../node/return.js";
import { ASSIGNERS } from "../../tokenize/rules.js";
import { RULE_FLAGS } from "../type.js";
import { COUNT_LOOP_RULES } from "./count-loop.js";
import { DICT_RULES } from "./dict.js";
import { LIST_LOOP_RULES } from "./list-loop.js";
import { PYTHON_COMPAT_RULES } from "./python-compat.js";
export const DOT_MEMBER_FUNCTION_INVOKE_RULES = [
  {
    pattern: [
      {
        type: Evaluable
      },
      {
        type: Expression,
        value: '.'
      },
      {
        type: FunctionInvoke
      }
    ],
    factory: (nodes, tokens)=>{
      return new MemberFunctionInvoke(nodes[0], nodes[2], tokens);
    }
  },
  {
    pattern: [
      {
        type: Evaluable
      },
      {
        type: Expression,
        value: '.'
      },
      {
        type: ValueWithParenthesis
      }
    ],
    factory: (nodes, tokens)=>{
      const wrapped = nodes[2];
      if (!(wrapped.value instanceof FunctionInvoke)) {
        return null;
      }
      return new MemberFunctionInvoke(nodes[0], wrapped.value, tokens);
    }
  },
  {
    pattern: [
      {
        type: Evaluable
      },
      {
        type: Expression,
        value: '.'
      },
      {
        type: IndexFetch
      }
    ],
    factory: (nodes, tokens)=>{
      const indexed = nodes[2];
      if (!(indexed.list instanceof FunctionInvoke)) {
        return null;
      }
      const memberInvoke = new MemberFunctionInvoke(nodes[0], indexed.list, tokens);
      return new IndexFetch(memberInvoke, indexed.index, tokens);
    }
  }
];
export const DOT_FETCH_MEMBER_RULES = [
  {
    pattern: [
      {
        type: Formula
      },
      {
        type: Expression,
        value: '.'
      },
      {
        type: Identifier
      }
    ],
    factory: (nodes, tokens)=>{
      if (nodes[2].value === '람다') {
        return null;
      }
      const formula = nodes[0];
      const lastTerm = formula.terms[formula.terms.length - 1];
      if (!lastTerm || typeof lastTerm !== 'object' || !Array.isArray(lastTerm.tokens)) {
        return null;
      }
      const fetched = new FetchMember(lastTerm, nodes[2].value, [
        ...lastTerm.tokens,
        ...nodes[1].tokens,
        ...nodes[2].tokens
      ]);
      return new Formula([
        ...formula.terms.slice(0, -1),
        fetched
      ], tokens);
    }
  },
  {
    pattern: [
      {
        type: Evaluable
      },
      {
        type: Expression,
        value: '.'
      },
      {
        type: Identifier
      }
    ],
    factory: (nodes, tokens)=>{
      if (nodes[2].value === '람다') {
        return null;
      }
      const target = nodes[0];
      const memberName = nodes[2].value;
      const terms = target instanceof Formula ? target.terms : target.terms;
      if (Array.isArray(terms) && terms.length > 0) {
        const lastTerm = terms[terms.length - 1];
        if (lastTerm && typeof lastTerm === 'object' && Array.isArray(lastTerm.tokens)) {
          const fetched = new FetchMember(lastTerm, memberName, [
            ...lastTerm.tokens,
            ...nodes[1].tokens,
            ...nodes[2].tokens
          ]);
          return new Formula([
            ...terms.slice(0, -1),
            fetched
          ], tokens);
        }
      }
      return new FetchMember(target, memberName, tokens);
    }
  }
];
export const DOT_ACCESS_RULES = [
  ...DOT_MEMBER_FUNCTION_INVOKE_RULES,
  ...DOT_FETCH_MEMBER_RULES
];
export const BASIC_RULES = [
  [
    // Template literal rules - must be processed early
    {
      pattern: [
        {
          type: TemplateStringPart
        },
        {
          type: Expression,
          value: '{'
        },
        {
          type: Evaluable
        },
        {
          type: Expression,
          value: '}'
        },
        {
          type: TemplateStringPart
        }
      ],
      factory: (nodes, tokens)=>{
        const startPart = nodes[0];
        const expr = nodes[2];
        const endPart = nodes[4];
        return new TemplateLiteral([
          startPart,
          expr,
          endPart
        ], tokens);
      }
    },
    {
      pattern: [
        {
          type: TemplateLiteral
        },
        {
          type: Expression,
          value: '{'
        },
        {
          type: Evaluable
        },
        {
          type: Expression,
          value: '}'
        },
        {
          type: TemplateStringPart
        }
      ],
      factory: (nodes, tokens)=>{
        const template = nodes[0];
        const expr = nodes[2];
        const endPart = nodes[4];
        return new TemplateLiteral([
          ...template.parts,
          expr,
          endPart
        ], tokens);
      }
    },
    ...[
      '참',
      '맞음',
      'True',
      'true'
    ].map((keyword)=>({
        pattern: [
          {
            type: Identifier,
            value: keyword
          }
        ],
        factory: (_nodes, tokens)=>{
          return new BooleanLiteral(true, tokens);
        }
      })),
    ...[
      '거짓',
      '아님',
      'False',
      'false'
    ].map((keyword)=>({
        pattern: [
          {
            type: Identifier,
            value: keyword
          }
        ],
        factory: (_nodes, tokens)=>{
          return new BooleanLiteral(false, tokens);
        }
      })),
    {
      pattern: [
        {
          type: EOL
        },
        {
          type: EOL
        }
      ],
      factory: (nodes, tokens)=>{
        const eol = nodes[0];
        eol.tokens = tokens;
        return eol;
      }
    },
    {
      pattern: [
        {
          type: Expression,
          value: ','
        },
        {
          type: EOL
        }
      ],
      factory: (nodes, tokens)=>{
        const comma = nodes[0];
        comma.tokens = tokens;
        return comma;
      }
    },
    {
      pattern: [
        {
          type: Evaluable
        },
        {
          type: Expression,
          value: '['
        },
        {
          type: Evaluable
        },
        {
          type: Expression,
          value: ']'
        }
      ],
      factory: (nodes, tokens)=>{
        const target = nodes[0];
        const index = nodes[2];
        return new IndexFetch(target, index, tokens);
      }
    }
  ],
  [
    {
      pattern: [
        {
          type: Expression,
          value: '('
        },
        {
          type: Expression,
          value: ')'
        }
      ],
      factory: (_nodes, tokens)=>new TupleLiteral([], tokens)
    },
    {
      pattern: [
        {
          type: Expression,
          value: '('
        },
        {
          type: Sequence
        },
        {
          type: Expression,
          value: ')'
        }
      ],
      factory: (nodes, tokens)=>{
        const sequence = nodes[1];
        return new TupleLiteral(sequence.items, tokens);
      }
    },
    {
      pattern: [
        {
          type: Expression,
          value: '('
        },
        {
          type: Evaluable
        },
        {
          type: Expression,
          value: ','
        },
        {
          type: Expression,
          value: ')'
        }
      ],
      factory: (nodes, tokens)=>{
        const item = nodes[1];
        return new TupleLiteral([
          item
        ], tokens);
      }
    },
    {
      pattern: [
        {
          type: Expression,
          value: '('
        },
        {
          type: Evaluable
        },
        {
          type: Expression,
          value: ')'
        }
      ],
      factory: (nodes, tokens)=>{
        const newNode = new ValueWithParenthesis(nodes[1], tokens);
        newNode.position = nodes[0].position;
        return newNode;
      }
    },
    {
      pattern: [
        {
          type: Expression,
          value: '['
        },
        {
          type: Sequence
        },
        {
          type: Expression,
          value: ']'
        }
      ],
      factory: (nodes, tokens)=>{
        const sequence = nodes[1];
        return new ListLiteral(sequence.items, tokens);
      }
    },
    {
      pattern: [
        {
          type: Evaluable
        },
        {
          type: Expression,
          value: '.'
        },
        {
          type: Identifier
        },
        {
          type: Operator
        },
        {
          type: Evaluable
        }
      ],
      factory: (nodes, tokens)=>{
        const left = new FetchMember(nodes[0], nodes[2].value, [
          ...nodes[0].tokens,
          ...nodes[1].tokens,
          ...nodes[2].tokens
        ]);
        const operator = nodes[3];
        const right = nodes[4];
        return new Formula([
          left,
          operator,
          right
        ], tokens);
      }
    },
    {
      pattern: [
        {
          type: Evaluable
        },
        {
          type: Operator
        },
        {
          type: Evaluable
        },
        {
          type: Expression,
          value: '.'
        },
        {
          type: Identifier
        }
      ],
      factory: (nodes, tokens)=>{
        const left = nodes[0];
        const operator = nodes[1];
        const right = new FetchMember(nodes[2], nodes[4].value, [
          ...nodes[2].tokens,
          ...nodes[3].tokens,
          ...nodes[4].tokens
        ]);
        if (left instanceof Formula) {
          return new Formula([
            ...left.terms,
            operator,
            right
          ], tokens);
        }
        return new Formula([
          left,
          operator,
          right
        ], tokens);
      }
    },
    {
      pattern: [
        {
          type: Evaluable
        },
        {
          type: Operator
        },
        {
          type: Evaluable
        }
      ],
      factory: (nodes, tokens)=>{
        const left = nodes[0];
        const operator = nodes[1];
        const right = nodes[2];
        if (left instanceof Formula) {
          return new Formula([
            ...left.terms,
            operator,
            right
          ], tokens);
        }
        return new Formula([
          left,
          operator,
          right
        ], tokens);
      }
    },
    {
      pattern: [
        {
          type: Operator,
          value: '!='
        }
      ],
      factory: (_nodes, tokens)=>new NotEqualOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '=='
        }
      ],
      factory: (_nodes, tokens)=>new EqualOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '>'
        }
      ],
      factory: (_nodes, tokens)=>new GreaterThanOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '<'
        }
      ],
      factory: (_nodes, tokens)=>new LessThanOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '>='
        }
      ],
      factory: (_nodes, tokens)=>new GreaterThanOrEqualOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '<='
        }
      ],
      factory: (_nodes, tokens)=>new LessThanOrEqualOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '//'
        }
      ],
      factory: (_nodes, tokens)=>new IntegerDivideOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '%'
        }
      ],
      factory: (_nodes, tokens)=>new ModularOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '**'
        }
      ],
      factory: (_nodes, tokens)=>new PowerOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '/'
        }
      ],
      factory: (_nodes, tokens)=>new DivideOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '*'
        }
      ],
      factory: (_nodes, tokens)=>new MultiplyOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '+'
        }
      ],
      factory: (nodes, tokens)=>new PlusOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '-'
        }
      ],
      factory: (nodes, tokens)=>new MinusOperator(tokens)
    },
    {
      pattern: [
        {
          type: Identifier,
          value: '이고'
        }
      ],
      factory: (_nodes, tokens)=>new AndOperator(tokens)
    },
    {
      pattern: [
        {
          type: Identifier,
          value: '고'
        }
      ],
      factory: (_nodes, tokens)=>new AndOperator(tokens)
    },
    {
      pattern: [
        {
          type: Identifier,
          value: '이거나'
        }
      ],
      factory: (_nodes, tokens)=>new OrOperator(tokens)
    },
    {
      pattern: [
        {
          type: Identifier,
          value: '거나'
        }
      ],
      factory: (_nodes, tokens)=>new OrOperator(tokens)
    },
    {
      pattern: [
        {
          type: Operator,
          value: '~'
        }
      ],
      factory: (_nodes, tokens)=>new RangeOperator(tokens)
    },
    {
      pattern: [
        {
          type: Expression,
          value: '!'
        },
        {
          type: Evaluable
        }
      ],
      factory: (nodes, tokens)=>{
        const evaluable = nodes[1];
        return new NotExpression(evaluable, tokens);
      }
    },
    {
      pattern: [
        {
          type: Evaluable
        },
        {
          type: Identifier,
          value: '아니다'
        }
      ],
      factory: (nodes, tokens)=>{
        const evaluable = nodes[0];
        return new NotExpression(evaluable, tokens);
      }
    }
  ]
];
export const ADVANCED_RULES = [
  {
    pattern: [
      {
        type: Identifier,
        value: '잠깐'
      },
      {
        type: Identifier,
        value: '멈추기'
      }
    ],
    factory: (_nodes, tokens)=>new Pause(tokens),
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: NewInstance
      },
      {
        type: TupleLiteral
      }
    ],
    factory: (nodes, tokens)=>{
      const ni = nodes[0];
      const tuple = nodes[1];
      return new NewInstance(ni.className, tuple.items, tokens);
    }
  },
  {
    pattern: [
      {
        type: NewInstance
      },
      {
        type: ValueWithParenthesis
      }
    ],
    factory: (nodes, tokens)=>{
      const ni = nodes[0];
      const vwp = nodes[1];
      let args = [];
      if (vwp.value instanceof Sequence) {
        args = vwp.value.items;
      } else {
        args = [
          vwp.value
        ];
      }
      return new NewInstance(ni.className, args, tokens);
    }
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '새'
      },
      {
        type: Identifier
      }
    ],
    factory: (nodes, tokens)=>{
      const className = nodes[1].value;
      return new NewInstance(className, [], tokens);
    }
  },
  ...PYTHON_COMPAT_RULES,
  {
    pattern: [
      {
        type: Identifier,
        value: '람다'
      },
      {
        type: Identifier
      },
      {
        type: Expression,
        value: ':'
      },
      {
        type: Evaluable
      }
    ],
    factory: (nodes, tokens)=>{
      const paramName = nodes[1].value;
      const body = nodes[3];
      return new LambdaLiteral([
        paramName
      ], body, tokens);
    }
  },
  {
    pattern: [
      {
        type: Evaluable
      },
      {
        type: Expression,
        value: ','
      },
      {
        type: Evaluable
      }
    ],
    factory: (nodes, tokens)=>{
      const a = nodes[0];
      const b = nodes[2];
      if (a instanceof Identifier && (a.value === '클래스' || a.value === '약속')) {
        return null;
      }
      return new Sequence([
        a,
        b
      ], tokens);
    }
  },
  {
    pattern: [
      {
        type: Sequence
      },
      {
        type: Expression,
        value: ','
      },
      {
        type: Evaluable
      }
    ],
    factory: (nodes, tokens)=>{
      const a = nodes[0];
      const b = nodes[2];
      return new Sequence([
        ...a.items,
        b
      ], tokens);
    }
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '람다'
      },
      {
        type: Sequence
      },
      {
        type: Expression,
        value: ':'
      },
      {
        type: Evaluable
      }
    ],
    factory: (nodes, tokens)=>{
      const params = extractLambdaParamNames(nodes[1]);
      if (!params) return null;
      return new LambdaLiteral(params, nodes[3], tokens);
    }
  },
  {
    pattern: [
      {
        type: Expression,
        value: '['
      },
      {
        type: Expression,
        value: ']'
      }
    ],
    factory: (_nodes, tokens)=>new ListLiteral([], tokens)
  },
  ...ASSIGNERS.map((assigner)=>({
      pattern: [
        {
          type: FetchMember
        },
        {
          type: Expression,
          value: assigner
        },
        {
          type: Evaluable
        }
      ],
      factory: (nodes, tokens)=>{
        const fetchMember = nodes[0];
        const operator = nodes[1];
        const value = nodes[2];
        return new SetMember(fetchMember.target, fetchMember.memberName, value, operator.value, tokens);
      },
      flags: [
        RULE_FLAGS.IS_STATEMENT
      ]
    })),
  ...ASSIGNERS.map((assigner)=>({
      pattern: [
        {
          type: IndexFetch
        },
        {
          type: Expression,
          value: assigner
        },
        {
          type: Evaluable
        }
      ],
      factory: (nodes, tokens)=>{
        const target = nodes[0];
        const operator = nodes[1];
        const value = nodes[2];
        return new SetToIndex(target, value, operator.value, tokens);
      },
      flags: [
        RULE_FLAGS.IS_STATEMENT
      ]
    })),
  ...ASSIGNERS.map((assigner)=>({
      pattern: [
        {
          type: Identifier
        },
        {
          type: Expression,
          value: assigner
        },
        {
          type: Evaluable
        }
      ],
      factory: (nodes, tokens)=>{
        const name = nodes[0].value;
        const operator = nodes[1];
        const value = nodes[2];
        return new SetVariable(name, value, tokens, operator.value);
      },
      flags: [
        RULE_FLAGS.IS_STATEMENT
      ]
    })),
  {
    pattern: [
      {
        type: IfStatement
      },
      {
        type: EOL
      },
      {
        type: ElseIfStatement
      }
    ],
    factory: (nodes, tokens)=>{
      const [ifStatement, _, elseIfStatement] = nodes;
      const elseIfCase = elseIfStatement.elseIfCase;
      ifStatement.cases.push(elseIfCase);
      ifStatement.tokens = tokens;
      return ifStatement;
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: IfStatement
      },
      {
        type: EOL
      },
      {
        type: ElseStatement
      }
    ],
    factory: (nodes, tokens)=>{
      const [ifStatement, _, elseStatement] = nodes;
      const elseCase = {
        body: elseStatement.body
      };
      ifStatement.cases.push(elseCase);
      ifStatement.tokens = tokens;
      return ifStatement;
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '아니면'
      },
      {
        type: Identifier,
        value: '만약'
      },
      {
        type: Evaluable
      },
      {
        type: Identifier,
        value: '이면'
      },
      {
        type: EOL
      },
      {
        type: Block
      }
    ],
    factory: (nodes, tokens)=>{
      const condition = nodes[2];
      const body = nodes[5];
      return new ElseIfStatement({
        condition,
        body
      }, tokens);
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '아니면'
      },
      {
        type: EOL
      },
      {
        type: Block
      }
    ],
    factory: (nodes, tokens)=>{
      const body = nodes[2];
      return new ElseStatement(body, tokens);
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '만약'
      },
      {
        type: Evaluable
      },
      {
        type: Identifier,
        value: '이면'
      },
      {
        type: EOL
      },
      {
        type: Block
      }
    ],
    factory: (nodes, tokens)=>{
      const condition = nodes[1];
      const body = nodes[4];
      return new IfStatement([
        {
          condition,
          body
        }
      ], tokens);
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Evaluable
      },
      {
        type: Identifier,
        value: '의'
      },
      {
        type: Identifier,
        value: '값'
      },
      {
        type: Identifier,
        value: '종류'
      }
    ],
    factory: (nodes, tokens)=>{
      const value = nodes[0];
      return new TypeOf(value, tokens);
    }
  },
  ...createTypeCastRules(),
  {
    pattern: [
      {
        type: Evaluable
      },
      {
        type: Identifier,
        value: '보여주기'
      }
    ],
    factory: (nodes, tokens)=>{
      const value = nodes[0];
      return new Print(value, tokens);
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Evaluable
      },
      {
        type: Identifier,
        value: '반환하기'
      }
    ],
    factory: (nodes, tokens)=>{
      const value = nodes[0];
      return new ReturnStatement(tokens, value);
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '반환하기'
      }
    ],
    factory: (_nodes, tokens)=>{
      return new ReturnStatement(tokens);
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '약속'
      },
      {
        type: Identifier,
        value: '그만'
      }
    ],
    factory: (_nodes, tokens)=>{
      return new ReturnStatement(tokens);
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '반복'
      },
      {
        type: EOL
      },
      {
        type: Block
      }
    ],
    factory: (nodes, tokens)=>new Loop(nodes[2], tokens),
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '반복하기'
      },
      {
        type: EOL
      },
      {
        type: Block
      }
    ],
    factory: (nodes, tokens)=>new Loop(nodes[2], tokens),
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '반복'
      },
      {
        type: Identifier,
        value: '그만'
      }
    ],
    factory: (_nodes, tokens)=>new Break(tokens),
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Identifier,
        value: '다음'
      },
      {
        type: Identifier,
        value: '반복'
      }
    ],
    factory: (_nodes, tokens)=>new Continue(tokens),
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  ...LIST_LOOP_RULES,
  ...COUNT_LOOP_RULES,
  {
    pattern: [
      {
        type: Identifier,
        value: '반복'
      },
      {
        type: Evaluable
      },
      {
        type: Identifier,
        value: '동안'
      },
      {
        type: EOL
      },
      {
        type: Block
      }
    ],
    factory: (nodes, tokens)=>{
      const condition = nodes[1];
      const body = nodes[4];
      return new ConditionalLoop(condition, body, tokens);
    },
    flags: [
      RULE_FLAGS.IS_STATEMENT
    ]
  },
  {
    pattern: [
      {
        type: Expression,
        value: '['
      },
      {
        type: Evaluable
      },
      {
        type: Expression,
        value: ']'
      }
    ],
    factory: (nodes, tokens)=>{
      const item = nodes[1];
      return new ListLiteral([
        item
      ], tokens);
    }
  },
  ...DICT_RULES
];
function createTypeCastRules() {
  const particles = [
    '을',
    '를'
  ];
  const targetTypes = [
    {
      keywords: [
        '숫자로'
      ],
      target: '숫자',
      split: [
        [
          '숫자',
          '로'
        ]
      ]
    },
    {
      keywords: [
        '문자열로',
        '문자로'
      ],
      target: '문자열',
      split: [
        [
          '문자열',
          '로'
        ],
        [
          '문자',
          '로'
        ]
      ]
    },
    {
      keywords: [
        '참거짓으로',
        '불리언으로'
      ],
      target: '참거짓',
      split: [
        [
          '참거짓',
          '으로'
        ],
        [
          '불리언',
          '으로'
        ]
      ]
    }
  ];
  const rules = [];
  for (const particle of particles){
    for (const { keywords, target, split } of targetTypes){
      for (const keyword of keywords){
        rules.push({
          pattern: [
            {
              type: Evaluable
            },
            {
              type: Identifier,
              value: particle
            },
            {
              type: Identifier,
              value: keyword
            },
            {
              type: Identifier,
              value: '바꾸기'
            }
          ],
          factory: (nodes, tokens)=>{
            const value = nodes[0];
            return new TypeCast(value, target, tokens);
          }
        });
      }
      if (split) {
        for (const [head, tail] of split){
          rules.push({
            pattern: [
              {
                type: Evaluable
              },
              {
                type: Identifier,
                value: particle
              },
              {
                type: Identifier,
                value: head
              },
              {
                type: Identifier,
                value: tail
              },
              {
                type: Identifier,
                value: '바꾸기'
              }
            ],
            factory: (nodes, tokens)=>{
              const value = nodes[0];
              return new TypeCast(value, target, tokens);
            }
          });
        }
      }
    }
  }
  return rules;
}
function extractLambdaParamNames(sequence) {
  const names = [];
  for (const item of sequence.items){
    if (!(item instanceof Identifier)) {
      return null;
    }
    names.push(item.value);
  }
  if (names.length === 0) {
    return null;
  }
  return names;
}
